{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9a7db64-5615-462a-adc3-446da0767246",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "from subprocess import call\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8edf64c6-81e8-412c-968d-115baeaa3be2",
   "metadata": {},
   "source": [
    "### Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error while creating /temp_vcf_out in project-GGy3Bb0JqBj7zfxY8v4by61X\r\n",
      "  The supplied authentication token has expired, code 401. Request Time=1677014317.689944, Request ID=unavailable\r\n",
      "Details: {\r\n",
      "    \"reason\": \"ExpiredToken\",\r\n",
      "    \"expired\": 1676670908898,\r\n",
      "    \"timeNow\": 1677014318469\r\n",
      "}\r\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\n",
    "number_of_threads = 4 # number of threads used in bcftools output compression\n",
    "tag_str = 'my_tag' # DNAnexus job tag\n",
    "\n",
    "project_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n",
    "\n",
    "# input vcf path (end with '/')\n",
    "dx_vcf_path = project_path + \"Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/\"\n",
    "# output vcf path (end with '/')\n",
    "dx_vcf_out_path = project_path + \"temp_vcf_out/\"\n",
    "# resource path (end with '/')\n",
    "dx_resource_path = project_path + \"GRCh38_resources/\"\n",
    "# difficult regions bed filename\n",
    "diff_bed = 'GRCh38_alldifficultregions.bed.gz'\n",
    "# reference genome filename\n",
    "ref_genome = 'GRCh38_reference_genome.fa' # index filename is inferred\n",
    "\n",
    "!dx mkdir -p {dx_vcf_out_path} # create gene.vcf output folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4000040-9b97-4085-a63e-b5aec384499f",
   "metadata": {},
   "source": [
    "### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2a8f096b-5b91-43aa-b546-4eaec039c30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_UKB_vcfs(df_gene, df_blk):\n",
    "    vcf_prefix = 'ukb23157_'\n",
    "    vcf_suffix = '_v1.vcf.gz'\n",
    "    \n",
    "    vcf_files = []\n",
    "    for i in range(df_gene.shape[0]):\n",
    "        vcf_files = vcf_files + df_blk.loc[(df_blk['seqname'] == df_gene[i,'seqname']) & \n",
    "                                           # when the gene is completely contained inside the block\n",
    "                                           (((df_blk['start_pos'] <= df_gene[i,'exon_flank_start']) & (df_blk['end_pos'] >= df_gene[i,'exon_flank_end'])) |\n",
    "                                            # when the block is completely contained inside the gene\n",
    "                                            ((df_blk['start_pos'] >= df_gene[i,'exon_flank_start']) & (df_blk['end_pos'] <= df_gene[i,'exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the start of the block\n",
    "                                            (df_blk['start_pos'].between(df_gene[i,'exon_flank_start'], df_gene[i,'exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the end of the block\n",
    "                                            (df_blk['end_pos'].between(df_gene[i,'exon_flank_start'], df_gene[i,'exon_flank_end']))),\n",
    "                                           'chr_blk_str'].tolist()\n",
    "    vcf_files = set(vcf_prefix + x + vcf_suffix for x in set(vcf_files))\n",
    "    return(vcf_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed62c94-9b34-42e2-90eb-67a527b04894",
   "metadata": {},
   "source": [
    "### List of gene symbols as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8fbe087c-8e0b-4b71-9f81-d9fbada151d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['BRCA1', 'TWNK', 'TENT2', 'PRR26', 'AFDN', 'UPK3B', 'SELENOS', 'DENND11', 'APOB']\n",
    "genes = ['BRCA1', 'APOB', 'TWNK']\n",
    "genes = ['']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "214b5b7e-a208-4c88-bf56-19dbcf0034bd",
   "metadata": {},
   "source": [
    "### Load pVCF block coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "62a230e4-d95d-4084-a914-01abecc3ff92",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>chr</th>\n",
       "      <th>blk</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>seqname</th>\n",
       "      <th>chr_blk_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1218130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1218131</td>\n",
       "      <td>1426969</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1426970</td>\n",
       "      <td>1758871</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1758872</td>\n",
       "      <td>2514221</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2514222</td>\n",
       "      <td>3782130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>973</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>135552245</td>\n",
       "      <td>141897932</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>974</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>141897933</td>\n",
       "      <td>152168662</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>975</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>152168663</td>\n",
       "      <td>153788223</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>976</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>153788224</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>977</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "      <td>cY_b0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>977 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ind  chr  blk  start_pos    end_pos seqname chr_blk_str\n",
       "0      1    1    0          1    1218130    chr1       c1_b0\n",
       "1      2    1    1    1218131    1426969    chr1       c1_b1\n",
       "2      3    1    2    1426970    1758871    chr1       c1_b2\n",
       "3      4    1    3    1758872    2514221    chr1       c1_b3\n",
       "4      5    1    4    2514222    3782130    chr1       c1_b4\n",
       "..   ...  ...  ...        ...        ...     ...         ...\n",
       "972  973   23   20  135552245  141897932    chrX      cX_b20\n",
       "973  974   23   21  141897933  152168662    chrX      cX_b21\n",
       "974  975   23   22  152168663  153788223    chrX      cX_b22\n",
       "975  976   23   23  153788224  156040895    chrX      cX_b23\n",
       "976  977   24    0          1   57227415    chrY       cY_b0\n",
       "\n",
       "[977 rows x 7 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blk = pd.read_table(\"./resources/pvcf_blocks.txt\", sep = '\\t', names = ['ind', 'chr', 'blk', 'start_pos', 'end_pos'])\n",
    "df_blk['seqname'] = 'chr' + df_blk['chr'].map(str)\n",
    "df_blk.loc[df_blk['seqname'] == 'chr23', 'seqname'] = 'chrX'\n",
    "df_blk.loc[df_blk['seqname'] == 'chr24', 'seqname'] = 'chrY'\n",
    "df_blk['chr_blk_str'] = df_blk['seqname'].str.replace('chr', 'c') + '_b' + df_blk['blk'].map(str)\n",
    "df_blk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16fc2f8f-3742-4c4e-b01e-05d35069d52d",
   "metadata": {},
   "source": [
    "### Load MANE transcript coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7c71254d-0da0-4458-92dd-3fcdf52e9b7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Polars found a filename. Ensure you pass a path to the file instead of a python file object when possible for best performance.INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'transcript_id', 'transcript_type', 'transcript_name', 'tag', 'protein_id', 'db_xref', 'exon_number', 'exon_id']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        white-space: pre;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-top: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-bottom: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        line-height: 95%;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "<small>shape: (127, 7)</small>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>\n",
       "seqname\n",
       "</th>\n",
       "<th>\n",
       "start\n",
       "</th>\n",
       "<th>\n",
       "end\n",
       "</th>\n",
       "<th>\n",
       "gene_name\n",
       "</th>\n",
       "<th>\n",
       "exon_flank_start\n",
       "</th>\n",
       "<th>\n",
       "exon_flank_end\n",
       "</th>\n",
       "<th>\n",
       "region\n",
       "</th>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "cat\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21043864\n",
       "</td>\n",
       "<td>\n",
       "21044073\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21043859\n",
       "</td>\n",
       "<td>\n",
       "21044078\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21043859-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21043513\n",
       "</td>\n",
       "<td>\n",
       "21043551\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21043508\n",
       "</td>\n",
       "<td>\n",
       "21043556\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21043508-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21042361\n",
       "</td>\n",
       "<td>\n",
       "21042476\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21042356\n",
       "</td>\n",
       "<td>\n",
       "21042481\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21042356-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21040938\n",
       "</td>\n",
       "<td>\n",
       "21041083\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21040933\n",
       "</td>\n",
       "<td>\n",
       "21041088\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21040933-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21037958\n",
       "</td>\n",
       "<td>\n",
       "21038111\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21037953\n",
       "</td>\n",
       "<td>\n",
       "21038116\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21037953-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21037100\n",
       "</td>\n",
       "<td>\n",
       "21037255\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21037095\n",
       "</td>\n",
       "<td>\n",
       "21037260\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21037095-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21035584\n",
       "</td>\n",
       "<td>\n",
       "21035708\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21035579\n",
       "</td>\n",
       "<td>\n",
       "21035713\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21035579-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21034816\n",
       "</td>\n",
       "<td>\n",
       "21034901\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21034811\n",
       "</td>\n",
       "<td>\n",
       "21034906\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21034811-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21033299\n",
       "</td>\n",
       "<td>\n",
       "21033518\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21033294\n",
       "</td>\n",
       "<td>\n",
       "21033523\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21033294-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21032354\n",
       "</td>\n",
       "<td>\n",
       "21032581\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21032349\n",
       "</td>\n",
       "<td>\n",
       "21032586\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21032349-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21029898\n",
       "</td>\n",
       "<td>\n",
       "21030015\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21029893\n",
       "</td>\n",
       "<td>\n",
       "21030020\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21029893-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr2&quot;\n",
       "</td>\n",
       "<td>\n",
       "21029639\n",
       "</td>\n",
       "<td>\n",
       "21029785\n",
       "</td>\n",
       "<td>\n",
       "&quot;APOB&quot;\n",
       "</td>\n",
       "<td>\n",
       "21029634\n",
       "</td>\n",
       "<td>\n",
       "21029790\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr2:21029634-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43082404\n",
       "</td>\n",
       "<td>\n",
       "43082575\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43082399\n",
       "</td>\n",
       "<td>\n",
       "43082580\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43082399...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43076488\n",
       "</td>\n",
       "<td>\n",
       "43076614\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43076483\n",
       "</td>\n",
       "<td>\n",
       "43076619\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43076483...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43074331\n",
       "</td>\n",
       "<td>\n",
       "43074521\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43074326\n",
       "</td>\n",
       "<td>\n",
       "43074526\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43074326...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43070928\n",
       "</td>\n",
       "<td>\n",
       "43071238\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43070923\n",
       "</td>\n",
       "<td>\n",
       "43071243\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43070923...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43067608\n",
       "</td>\n",
       "<td>\n",
       "43067695\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43067603\n",
       "</td>\n",
       "<td>\n",
       "43067700\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43067603...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063874\n",
       "</td>\n",
       "<td>\n",
       "43063951\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063869\n",
       "</td>\n",
       "<td>\n",
       "43063956\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43063869...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063333\n",
       "</td>\n",
       "<td>\n",
       "43063373\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063328\n",
       "</td>\n",
       "<td>\n",
       "43063378\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43063328...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43057052\n",
       "</td>\n",
       "<td>\n",
       "43057135\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43057047\n",
       "</td>\n",
       "<td>\n",
       "43057140\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43057047...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43051063\n",
       "</td>\n",
       "<td>\n",
       "43051117\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43051058\n",
       "</td>\n",
       "<td>\n",
       "43051122\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43051058...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43049121\n",
       "</td>\n",
       "<td>\n",
       "43049194\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43049116\n",
       "</td>\n",
       "<td>\n",
       "43049199\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43049116...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43047643\n",
       "</td>\n",
       "<td>\n",
       "43047703\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43047638\n",
       "</td>\n",
       "<td>\n",
       "43047708\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43047638...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43044295\n",
       "</td>\n",
       "<td>\n",
       "43045802\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43044290\n",
       "</td>\n",
       "<td>\n",
       "43045807\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43044290...\n",
       "</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "shape: (127, 7)\n",
       "┌─────────┬──────────┬──────────┬───────────┬──────────────────┬────────────────┬──────────────────┐\n",
       "│ seqname ┆ start    ┆ end      ┆ gene_name ┆ exon_flank_start ┆ exon_flank_end ┆ region           │\n",
       "│ ---     ┆ ---      ┆ ---      ┆ ---       ┆ ---              ┆ ---            ┆ ---              │\n",
       "│ cat     ┆ i64      ┆ i64      ┆ str       ┆ i64              ┆ i64            ┆ str              │\n",
       "╞═════════╪══════════╪══════════╪═══════════╪══════════════════╪════════════════╪══════════════════╡\n",
       "│ chr2    ┆ 21043864 ┆ 21044073 ┆ APOB      ┆ 21043859         ┆ 21044078       ┆ chr2:21043859-21 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 044078           │\n",
       "│ chr2    ┆ 21043513 ┆ 21043551 ┆ APOB      ┆ 21043508         ┆ 21043556       ┆ chr2:21043508-21 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 043556           │\n",
       "│ chr2    ┆ 21042361 ┆ 21042476 ┆ APOB      ┆ 21042356         ┆ 21042481       ┆ chr2:21042356-21 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 042481           │\n",
       "│ chr2    ┆ 21040938 ┆ 21041083 ┆ APOB      ┆ 21040933         ┆ 21041088       ┆ chr2:21040933-21 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 041088           │\n",
       "│ ...     ┆ ...      ┆ ...      ┆ ...       ┆ ...              ┆ ...            ┆ ...              │\n",
       "│ chr17   ┆ 43051063 ┆ 43051117 ┆ BRCA1     ┆ 43051058         ┆ 43051122       ┆ chr17:43051058-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3051122          │\n",
       "│ chr17   ┆ 43049121 ┆ 43049194 ┆ BRCA1     ┆ 43049116         ┆ 43049199       ┆ chr17:43049116-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3049199          │\n",
       "│ chr17   ┆ 43047643 ┆ 43047703 ┆ BRCA1     ┆ 43047638         ┆ 43047708       ┆ chr17:43047638-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3047708          │\n",
       "│ chr17   ┆ 43044295 ┆ 43045802 ┆ BRCA1     ┆ 43044290         ┆ 43045807       ┆ chr17:43044290-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3045807          │\n",
       "└─────────┴──────────┴──────────┴───────────┴──────────────────┴────────────────┴──────────────────┘"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = read_gtf(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.gtf.gz\")\n",
    "df = df.filter(pl.col('feature') == 'exon')\n",
    "df = df.filter(pl.col('gene_name').is_in(genes))\n",
    "df = df.select(pl.col(['seqname', 'start', 'end', 'gene_name']))\n",
    "df = df.with_columns([(pl.col('start') - exon_flank_nt).alias('exon_flank_start'), \n",
    "                      (pl.col('end') + exon_flank_nt).alias('exon_flank_end')])\n",
    "df = df.with_column((pl.col('seqname')+':'+pl.col('exon_flank_start')+'-'+pl.col('exon_flank_end')).alias('region'))\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e66005-ff41-4b75-a443-a2ed9df136bd",
   "metadata": {},
   "source": [
    "### Run Swiss-army-knife on DNAnexus\n",
    "- get region info for each gene from block file\n",
    "- bcftools command for step 2 finished\n",
    "- bcftools command for step 3 & 4 TBD\n",
    "- list genes not included in MANE set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "14ba3303-3410-4c56-a1f4-c6201572c1be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GPGYKP8JqBj7XK92KqZ52Qpz\n",
      "\u001b[0mjob-GPGYKPQJqBjPbv6Q5X7JVB7G\n",
      "\u001b[0mjob-GPGYKPjJqBjJyXxpzJq3Xfv9\n",
      "\u001b[0mGenes not found in MANE database:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "genes_not_found = []\n",
    "genes_found = []\n",
    "\n",
    "for gene in genes:\n",
    "    df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "    if df_gene.shape[0] > 0:\n",
    "        genes_found.append(gene)\n",
    "        \n",
    "        vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "        vcf_str = ' '.join(vcf_files)\n",
    "        region_str = ','.join(df_gene['region'].to_list())\n",
    "        mem_level = len(vcf_files) # dynamically change memory level when submitting jobs on DNAnexus\n",
    "\n",
    "        bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "        bcftools_cmd2 = \"bcftools view -Ou -T ^\" + diff_bed\n",
    "        bcftools_cmd3 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "        bcftools_cmd4 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "        bcftools_cmd5 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "        bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "\n",
    "        dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "        dx_command = 'dx run swiss-army-knife --instance-type mem' + str(mem_level) + '_ssd1_v2_x4 -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "        !{dx_command}\n",
    "    else:\n",
    "        genes_not_found.append(gene)\n",
    "            \n",
    "print('Genes not found in MANE database:')\n",
    "print(genes_not_found)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e454b2db-5ba6-446f-8c3c-31c164c7df92",
   "metadata": {},
   "source": [
    "### Below is for testing purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4df214f4-69b6-4396-a167-e7a0ff599e41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\n",
      "dx run swiss-army-knife --instance-type mem2_ssd1_v2_x4 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/temp_vcf_out/ --tag \"my_tag\" --property gene=APOB\n"
     ]
    }
   ],
   "source": [
    "gene = 'APOB'\n",
    "df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "vcf_str = ' '.join(vcf_files)\n",
    "region_str = ','.join(df_gene['region'].to_list())\n",
    "mem_level = len(vcf_files)\n",
    "\n",
    "bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "bcftools_cmd2 = \"bcftools view -Ou -T ^\" + diff_bed\n",
    "bcftools_cmd3 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "bcftools_cmd4 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "bcftools_cmd5 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "\n",
    "dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "dx_command = 'dx run swiss-army-knife --instance-type mem' + str(mem_level) + '_ssd1_v2_x4 -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "\n",
    "print(bcftools_command)\n",
    "print(dx_command)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4d4d5547-7ac4-4a6b-8409-130c10019c9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GPG9KY8JqBj5gK412JXv3f8G\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!{dx_command}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
