{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9a7db64-5615-462a-adc3-446da0767246",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import numpy as np\n",
    "# import polars as pl\n",
    "import pandas as pd\n",
    "from subprocess import call\n",
    "# from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8edf64c6-81e8-412c-968d-115baeaa3be2",
   "metadata": {},
   "source": [
    "### Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\n",
    "number_of_threads = 4 # number of threads used in bcftools output compression\n",
    "tag_str = 'Tian_080423' # DNAnexus job tag\n",
    "\n",
    "project_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n",
    "\n",
    "# DNAnexus input vcf path (end with '/')\n",
    "dx_vcf_path = project_path + \"Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/\"\n",
    "# DNAnexus gene-specific vcf output path (end with '/')\n",
    "dx_vcf_out_path = project_path + \"Tian_folder/14gene_vcf/\"\n",
    "# DNAnexus heterogeneous carrier output path (under dx_vcf_out_path)\n",
    "dx_carrier_out_path = \"hetero_carrier/\"\n",
    "# DNAnexus resource path (end with '/')\n",
    "dx_resource_path = project_path + \"GRCh38_resources/\"\n",
    "# DNAnexus script path (end with '/')\n",
    "dx_script_path = project_path + \"scripts/\"\n",
    "# difficult regions bed file (under dx_resource_path)\n",
    "diff_bed = 'GRCh38_alldifficultregions.bed.gz'\n",
    "# reference genome file (under dx_resource_path)\n",
    "ref_genome = 'GRCh38_reference_genome.fa' # index filename is inferred\n",
    "\n",
    "!dx mkdir -p {dx_vcf_out_path} # create gene.vcf output folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4000040-9b97-4085-a63e-b5aec384499f",
   "metadata": {},
   "source": [
    "### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2a8f096b-5b91-43aa-b546-4eaec039c30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_UKB_vcfs(df_gene, df_blk):\n",
    "    vcf_prefix = 'ukb23157_'\n",
    "    vcf_suffix = '_v1.vcf.gz'\n",
    "    \n",
    "    vcf_files = []\n",
    "    for index, row in df_gene.iterrows():\n",
    "        vcf_files = vcf_files + df_blk.loc[(df_blk['seqname'] == row['seqname']) & \n",
    "                                           # when the gene is completely contained inside the block\n",
    "                                           (((df_blk['start_pos'] <= row['exon_flank_start']) & (df_blk['end_pos'] >= row['exon_flank_end'])) |\n",
    "                                            # when the block is completely contained inside the gene\n",
    "                                            ((df_blk['start_pos'] >= row['exon_flank_start']) & (df_blk['end_pos'] <= row['exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the start of the block\n",
    "                                            (df_blk['start_pos'].between(row['exon_flank_start'], row['exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the end of the block\n",
    "                                            (df_blk['end_pos'].between(row['exon_flank_start'], row['exon_flank_end']))),\n",
    "                                           'chr_blk_str'].tolist()\n",
    "    vcf_files = set(vcf_prefix + x + vcf_suffix for x in set(vcf_files))\n",
    "    return(vcf_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed62c94-9b34-42e2-90eb-67a527b04894",
   "metadata": {},
   "source": [
    "### List of gene symbols as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8fbe087c-8e0b-4b71-9f81-d9fbada151d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['LDLRAP1', 'PSMD4', 'GPN2', 'RABIF', 'RAB10', 'PREB', 'HMGCR', 'MYLIP', 'NPC1L1', 'ABCA1', 'HNF1A', 'OTX2', 'HNF4A', 'SOD1']\n",
    "\n",
    "# # to read from txt:\n",
    "# with open(\"resources/my_gene_file.txt\", \"r\") as gene_file:\n",
    "#     genes = gene_file.readlines()\n",
    "#     genes = [l.replace(\"\\n\", \"\") for l in genes]\n",
    "#     print(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "214b5b7e-a208-4c88-bf56-19dbcf0034bd",
   "metadata": {},
   "source": [
    "### Load pVCF block coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "62a230e4-d95d-4084-a914-01abecc3ff92",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>chr</th>\n",
       "      <th>blk</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>seqname</th>\n",
       "      <th>chr_blk_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1218130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1218131</td>\n",
       "      <td>1426969</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1426970</td>\n",
       "      <td>1758871</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1758872</td>\n",
       "      <td>2514221</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2514222</td>\n",
       "      <td>3782130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>973</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>135552245</td>\n",
       "      <td>141897932</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>974</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>141897933</td>\n",
       "      <td>152168662</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>975</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>152168663</td>\n",
       "      <td>153788223</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>976</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>153788224</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>977</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "      <td>cY_b0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>977 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ind  chr  blk  start_pos    end_pos seqname chr_blk_str\n",
       "0      1    1    0          1    1218130    chr1       c1_b0\n",
       "1      2    1    1    1218131    1426969    chr1       c1_b1\n",
       "2      3    1    2    1426970    1758871    chr1       c1_b2\n",
       "3      4    1    3    1758872    2514221    chr1       c1_b3\n",
       "4      5    1    4    2514222    3782130    chr1       c1_b4\n",
       "..   ...  ...  ...        ...        ...     ...         ...\n",
       "972  973   23   20  135552245  141897932    chrX      cX_b20\n",
       "973  974   23   21  141897933  152168662    chrX      cX_b21\n",
       "974  975   23   22  152168663  153788223    chrX      cX_b22\n",
       "975  976   23   23  153788224  156040895    chrX      cX_b23\n",
       "976  977   24    0          1   57227415    chrY       cY_b0\n",
       "\n",
       "[977 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blk = pd.read_table(\"./resources/pvcf_blocks.txt\", sep = '\\t', names = ['ind', 'chr', 'blk', 'start_pos', 'end_pos'])\n",
    "df_blk['seqname'] = 'chr' + df_blk['chr'].map(str)\n",
    "df_blk.loc[df_blk['seqname'] == 'chr23', 'seqname'] = 'chrX'\n",
    "df_blk.loc[df_blk['seqname'] == 'chr24', 'seqname'] = 'chrY'\n",
    "df_blk['chr_blk_str'] = df_blk['seqname'].str.replace('chr', 'c') + '_b' + df_blk['blk'].map(str)\n",
    "df_blk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16fc2f8f-3742-4c4e-b01e-05d35069d52d",
   "metadata": {},
   "source": [
    "### Load MANE transcript coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c71254d-0da0-4458-92dd-3fcdf52e9b7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_flank_start</th>\n",
       "      <th>exon_flank_end</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25543606</td>\n",
       "      <td>25543786</td>\n",
       "      <td>LDLRAP1</td>\n",
       "      <td>25543601</td>\n",
       "      <td>25543791</td>\n",
       "      <td>chr1:25543601-25543791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1479</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25553922</td>\n",
       "      <td>25554064</td>\n",
       "      <td>LDLRAP1</td>\n",
       "      <td>25553917</td>\n",
       "      <td>25554069</td>\n",
       "      <td>chr1:25553917-25554069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1481</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25554860</td>\n",
       "      <td>25554972</td>\n",
       "      <td>LDLRAP1</td>\n",
       "      <td>25554855</td>\n",
       "      <td>25554977</td>\n",
       "      <td>chr1:25554855-25554977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1483</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25557153</td>\n",
       "      <td>25557267</td>\n",
       "      <td>LDLRAP1</td>\n",
       "      <td>25557148</td>\n",
       "      <td>25557272</td>\n",
       "      <td>chr1:25557148-25557272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1485</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25562644</td>\n",
       "      <td>25562716</td>\n",
       "      <td>LDLRAP1</td>\n",
       "      <td>25562639</td>\n",
       "      <td>25562721</td>\n",
       "      <td>chr1:25562639-25562721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486568</th>\n",
       "      <td>chr21</td>\n",
       "      <td>31659693</td>\n",
       "      <td>31659841</td>\n",
       "      <td>SOD1</td>\n",
       "      <td>31659688</td>\n",
       "      <td>31659846</td>\n",
       "      <td>chr21:31659688-31659846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486571</th>\n",
       "      <td>chr21</td>\n",
       "      <td>31663790</td>\n",
       "      <td>31663886</td>\n",
       "      <td>SOD1</td>\n",
       "      <td>31663785</td>\n",
       "      <td>31663891</td>\n",
       "      <td>chr21:31663785-31663891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486573</th>\n",
       "      <td>chr21</td>\n",
       "      <td>31666449</td>\n",
       "      <td>31666518</td>\n",
       "      <td>SOD1</td>\n",
       "      <td>31666444</td>\n",
       "      <td>31666523</td>\n",
       "      <td>chr21:31666444-31666523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486575</th>\n",
       "      <td>chr21</td>\n",
       "      <td>31667258</td>\n",
       "      <td>31667375</td>\n",
       "      <td>SOD1</td>\n",
       "      <td>31667253</td>\n",
       "      <td>31667380</td>\n",
       "      <td>chr21:31667253-31667380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486577</th>\n",
       "      <td>chr21</td>\n",
       "      <td>31668471</td>\n",
       "      <td>31668931</td>\n",
       "      <td>SOD1</td>\n",
       "      <td>31668466</td>\n",
       "      <td>31668936</td>\n",
       "      <td>chr21:31668466-31668936</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>167 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname     start       end gene_name  exon_flank_start   \n",
       "1476      chr1  25543606  25543786   LDLRAP1          25543601  \\\n",
       "1479      chr1  25553922  25554064   LDLRAP1          25553917   \n",
       "1481      chr1  25554860  25554972   LDLRAP1          25554855   \n",
       "1483      chr1  25557153  25557267   LDLRAP1          25557148   \n",
       "1485      chr1  25562644  25562716   LDLRAP1          25562639   \n",
       "...        ...       ...       ...       ...               ...   \n",
       "486568   chr21  31659693  31659841      SOD1          31659688   \n",
       "486571   chr21  31663790  31663886      SOD1          31663785   \n",
       "486573   chr21  31666449  31666518      SOD1          31666444   \n",
       "486575   chr21  31667258  31667375      SOD1          31667253   \n",
       "486577   chr21  31668471  31668931      SOD1          31668466   \n",
       "\n",
       "        exon_flank_end                   region  \n",
       "1476          25543791   chr1:25543601-25543791  \n",
       "1479          25554069   chr1:25553917-25554069  \n",
       "1481          25554977   chr1:25554855-25554977  \n",
       "1483          25557272   chr1:25557148-25557272  \n",
       "1485          25562721   chr1:25562639-25562721  \n",
       "...                ...                      ...  \n",
       "486568        31659846  chr21:31659688-31659846  \n",
       "486571        31663891  chr21:31663785-31663891  \n",
       "486573        31666523  chr21:31666444-31666523  \n",
       "486575        31667380  chr21:31667253-31667380  \n",
       "486577        31668936  chr21:31668466-31668936  \n",
       "\n",
       "[167 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = read_gtf(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.gtf.gz\")\n",
    "# df = df.filter(pl.col('feature') == 'exon')\n",
    "# df = df.filter(pl.col('gene_name').is_in(genes))\n",
    "# df = df.select(pl.col(['seqname', 'start', 'end', 'gene_name']))\n",
    "# df = df.with_columns([(pl.col('start') - exon_flank_nt).alias('exon_flank_start'), \n",
    "#                       (pl.col('end') + exon_flank_nt).alias('exon_flank_end')])\n",
    "# df = df.with_column((pl.col('seqname')+':'+pl.col('exon_flank_start')+'-'+pl.col('exon_flank_end')).alias('region'))\n",
    "# df\n",
    "\n",
    "df = pd.read_csv(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz\")\n",
    "df = df.loc[(df['feature'] == 'exon') & (df['gene_name'].isin(genes))]\n",
    "df = df[['seqname', 'start', 'end', 'gene_name']]\n",
    "df['exon_flank_start'] = df['start'] - exon_flank_nt\n",
    "df['exon_flank_end'] = df['end'] + exon_flank_nt\n",
    "df['region'] = ((df['seqname'] + ':').str.cat(df['exon_flank_start'].astype(str)) + '-').str.cat(df['exon_flank_end'].astype(str))\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e66005-ff41-4b75-a443-a2ed9df136bd",
   "metadata": {},
   "source": [
    "### Run Swiss-army-knife on DNAnexus\n",
    "- get region info for each gene from block file\n",
    "- bcftools command for step 2 finished\n",
    "- bcftools command for step 3 & 4 TBD\n",
    "- list genes not included in MANE set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "14ba3303-3410-4c56-a1f4-c6201572c1be",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GY6Yx88JqBj16f1YBv2JJP99\n",
      "\u001b[0mjob-GY6Yx8QJqBjFvP5vf4K7qZbv\n",
      "\u001b[0mjob-GY6Yx8jJqBj61fZpBGXV8jbZ\n",
      "\u001b[0mjob-GY6Yx98JqBj3FjfKbq7fqFK6\n",
      "\u001b[0mjob-GY6Yx9QJqBj7zQz8xQFbk43Q\n",
      "\u001b[0mjob-GY6Yx9jJqBjJpJ1YV1ppZy8p\n",
      "\u001b[0mjob-GY6YxBQJqBj9031xQk349Kq6\n",
      "\u001b[0mjob-GY6YxF0JqBj02FPzf62PKpj2\n",
      "\u001b[0mjob-GY6YxF8JqBj1P949B6Y2kZ1p\n",
      "\u001b[0mjob-GY6YxFQJqBjF697XkKbZFf9P\n",
      "\u001b[0mjob-GY6YxFjJqBj23473kKQZ04kZ\n",
      "\u001b[0mjob-GY6YxG8JqBj29bkz9zXgpQKF\n",
      "\u001b[0mjob-GY6YxGQJqBjFvb3BvFbgXzzV\n",
      "\u001b[0mjob-GY6YxGjJqBj11qkkV97xbv25\n",
      "\u001b[0mGenes not found in MANE database:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "genes_not_found = []\n",
    "genes_found = []\n",
    "known_large_genes = [\"DSP\", \"TSC2\", \"TTN\", \"NCOA3\"]\n",
    "\n",
    "for gene in genes:\n",
    "#     df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "    df_gene = df.loc[df['gene_name'] == gene]\n",
    "    if df_gene.shape[0] > 0:\n",
    "        genes_found.append(gene)\n",
    "        \n",
    "        vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "        vcf_str = ' '.join(vcf_files)\n",
    "        region_str = ','.join(df_gene['region'].to_list())\n",
    "        \n",
    "        if ((len(vcf_files) > 1) or (gene in known_large_genes)):\n",
    "            mem_level = \"mem2_ssd1_v2_x16\" # dynamically change memory level when submitting jobs on DNAnexus\n",
    "        else:\n",
    "            mem_level = \"mem1_ssd1_v2_x4\"\n",
    "        \n",
    "        # filtering commands\n",
    "        bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "        bcftools_cmd2 = \"bcftools view -Ou --max-alleles 5 -T ^\" + diff_bed\n",
    "        bcftools_cmd3 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "        bcftools_cmd4 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "        bcftools_cmd5 = \"bcftools view -Oz -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \"_variants.vcf.gz\"\n",
    "        \n",
    "        # variant-only vcf command \n",
    "        bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "        \n",
    "        # get genotypes\n",
    "        genotype_cmd0 = \"mkdir -p \" + dx_carrier_out_path\n",
    "        genotype_cmd1 = \"bcftools query -i 'GT=\\\\\\\"RA\\\\\\\"|GT=\\\\\\\"AR\\\\\\\"' -f '%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n' \" + gene + \"_variants.vcf.gz > \" + dx_carrier_out_path + gene + \".ssv\"\n",
    "        \n",
    "        # get genotypes command\n",
    "        genotype_command = \" && \".join([genotype_cmd0, genotype_cmd1])\n",
    "\n",
    "        dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "        dx_command = 'dx run swiss-army-knife --instance-type ' + mem_level + ' -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + ' && ' + genotype_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "\n",
    "        !{dx_command}\n",
    "    else:\n",
    "        genes_not_found.append(gene)\n",
    "            \n",
    "print('Genes not found in MANE database:')\n",
    "print(genes_not_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "adf05525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dx run swiss-army-knife --instance-type mem1_ssd1_v2_x4 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c1_b36_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c1_b36_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr1:55039543-55040049,chr1:55043838-55044039,chr1:55046518-55046651,chr1:55052273-55052416,chr1:55052645-55052796,chr1:55055988-55056194,chr1:55057326-55057519,chr1:55058031-55058214,chr1:55058494-55058652,chr1:55059481-55059668,chr1:55061370-55061561,chr1:55063364-55064857 ukb23157_c1_b36_v1.vcf.gz | bcftools view -Ou --max-alleles 5 -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id \\'%CHROM\\\\_%POS\\\\_%REF\\\\_%ALT\\' | bcftools view -Oz -i \\'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\"mis\\\\\")> 0.9\\' > PCSK9_variants.vcf.gz && mkdir -p hetero_carrier/ && bcftools query -i \\'GT=\\\\\"RA\\\\\"|GT=\\\\\"AR\\\\\"\\' -f \\'%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n\\' PCSK9_variants.vcf.gz > hetero_carrier/PCSK9.ssv\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/Vineel_genes/filtered_vcf_5/ --tag \"Vineel_070722\" --property gene=PCSK9'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dx_command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7c48709",
   "metadata": {},
   "outputs": [],
   "source": [
    "dx run swiss-army-knife \n",
    "--instance-type mem1_ssd1_v2_x4 \n",
    "-y --brief \n",
    "-iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c1_b36_v1.vcf.gz\" \n",
    "-iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c1_b36_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" \n",
    "-iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" \n",
    "-iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" \n",
    "-icmd=\"bcftools concat -Ou -a -r chr1:55039543-55040049,chr1:55043838-55044039,chr1:55046518-55046651,chr1:55052273-55052416,chr1:55052645-55052796,chr1:55055988-55056194,chr1:55057326-55057519,chr1:55058031-55058214,chr1:55058494-55058652,chr1:55059481-55059668,chr1:55061370-55061561,chr1:55063364-55064857 ukb23157_c1_b36_v1.vcf.gz | bcftools view -Ou --max-alleles 5 -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id \\'%CHROM\\\\_%POS\\\\_%REF\\\\_%ALT\\' | bcftools view -Oz -i \\'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\"mis\\\\\")> 0.9\\' > PCSK9_variants.vcf.gz && mkdir -p hetero_carrier/ && bcftools query -i \\'GT=\\\\\"RA\\\\\"|GT=\\\\\"AR\\\\\"\\' -f \\'%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n\\' PCSK9_variants.vcf.gz > hetero_carrier/PCSK9.ssv\" \n",
    "--destination project-GGy3Bb0JqBj7zfxY8v4by61X:/Vineel_genes/filtered_vcf_5/ \n",
    "--tag \"Vineel_070722\"\n",
    "--property gene=PCSK9"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e454b2db-5ba6-446f-8c3c-31c164c7df92",
   "metadata": {},
   "source": [
    "### Below is for testing purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4df214f4-69b6-4396-a167-e7a0ff599e41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\n",
      "dx run swiss-army-knife --instance-type mem2_ssd1_v2_x4 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/temp_vcf_out/ --tag \"my_tag\" --property gene=APOB\n"
     ]
    }
   ],
   "source": [
    "gene = 'APOB'\n",
    "df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "vcf_str = ' '.join(vcf_files)\n",
    "region_str = ','.join(df_gene['region'].to_list())\n",
    "mem_level = len(vcf_files)\n",
    "\n",
    "bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "bcftools_cmd2 = \"bcftools view -Ou -T ^\" + diff_bed\n",
    "bcftools_cmd3 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "bcftools_cmd4 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "bcftools_cmd5 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "\n",
    "dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "dx_command = 'dx run swiss-army-knife --instance-type mem' + str(mem_level) + '_ssd1_v2_x4 -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "\n",
    "print(bcftools_command)\n",
    "print(dx_command)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4d4d5547-7ac4-4a6b-8409-130c10019c9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GPG9KY8JqBj5gK412JXv3f8G\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!{dx_command}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
