{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a9a7db64-5615-462a-adc3-446da0767246",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "from subprocess import call\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8edf64c6-81e8-412c-968d-115baeaa3be2",
   "metadata": {},
   "source": [
    "### Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\n",
    "number_of_threads = 4 # number of threads used in bcftools output compression\n",
    "tag_str = 'Vineel_090823' # DNAnexus job tag\n",
    "\n",
    "project_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n",
    "\n",
    "# input vcf path (end with '/')\n",
    "dx_vcf_path = project_path + \"Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/\"\n",
    "# output vcf path (end with '/')\n",
    "dx_vcf_out_path = project_path + \"Tian_folder/Vineel_genes_test/\"\n",
    "# resource path (end with '/')\n",
    "dx_resource_path = project_path + \"GRCh38_resources/\"\n",
    "# difficult regions bed filename\n",
    "diff_bed = 'GRCh38_alldifficultregions.bed.gz'\n",
    "# reference genome filename\n",
    "ref_genome = 'GRCh38_reference_genome.fa' # index filename is inferred\n",
    "\n",
    "!dx mkdir -p {dx_vcf_out_path} # create gene.vcf output folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4000040-9b97-4085-a63e-b5aec384499f",
   "metadata": {},
   "source": [
    "### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2a8f096b-5b91-43aa-b546-4eaec039c30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_UKB_vcfs(df_gene, df_blk):\n",
    "    vcf_prefix = 'ukb23157_'\n",
    "    vcf_suffix = '_v1.vcf.gz'\n",
    "    \n",
    "    vcf_files = []\n",
    "    for index, row in df_gene.iterrows():\n",
    "        vcf_files = vcf_files + df_blk.loc[(df_blk['seqname'] == row['seqname']) & \n",
    "                                           # when the gene is completely contained inside the block\n",
    "                                           (((df_blk['start_pos'] <= row['exon_flank_start']) & (df_blk['end_pos'] >= row['exon_flank_end'])) |\n",
    "                                            # when the block is completely contained inside the gene\n",
    "                                            ((df_blk['start_pos'] >= row['exon_flank_start']) & (df_blk['end_pos'] <= row['exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the start of the block\n",
    "                                            (df_blk['start_pos'].between(row['exon_flank_start'], row['exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the end of the block\n",
    "                                            (df_blk['end_pos'].between(row['exon_flank_start'], row['exon_flank_end']))),\n",
    "                                           'chr_blk_str'].tolist()\n",
    "    vcf_files = set(vcf_prefix + x + vcf_suffix for x in set(vcf_files))\n",
    "    return(vcf_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed62c94-9b34-42e2-90eb-67a527b04894",
   "metadata": {},
   "source": [
    "### List of gene symbols as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8fbe087c-8e0b-4b71-9f81-d9fbada151d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes = [\"BRCA1\", \"BRCA2\", \"MSH2\", \"MSH6\", \"MLH1\", \"PMS2\", \"EPCAM\"]\n",
    "genes = [\"LPA\"]\n",
    "\n",
    "# # to read from txt:\n",
    "# with open(\"resources/my_gene_file.txt\", \"r\") as gene_file:\n",
    "#     genes = gene_file.readlines()\n",
    "#     genes = [l.replace(\"\\n\", \"\") for l in genes]\n",
    "#     print(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "214b5b7e-a208-4c88-bf56-19dbcf0034bd",
   "metadata": {},
   "source": [
    "### Load pVCF block coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "62a230e4-d95d-4084-a914-01abecc3ff92",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>chr</th>\n",
       "      <th>blk</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>seqname</th>\n",
       "      <th>chr_blk_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1218130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1218131</td>\n",
       "      <td>1426969</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1426970</td>\n",
       "      <td>1758871</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1758872</td>\n",
       "      <td>2514221</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2514222</td>\n",
       "      <td>3782130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>973</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>135552245</td>\n",
       "      <td>141897932</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>974</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>141897933</td>\n",
       "      <td>152168662</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>975</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>152168663</td>\n",
       "      <td>153788223</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>976</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>153788224</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>977</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "      <td>cY_b0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>977 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ind  chr  blk  start_pos    end_pos seqname chr_blk_str\n",
       "0      1    1    0          1    1218130    chr1       c1_b0\n",
       "1      2    1    1    1218131    1426969    chr1       c1_b1\n",
       "2      3    1    2    1426970    1758871    chr1       c1_b2\n",
       "3      4    1    3    1758872    2514221    chr1       c1_b3\n",
       "4      5    1    4    2514222    3782130    chr1       c1_b4\n",
       "..   ...  ...  ...        ...        ...     ...         ...\n",
       "972  973   23   20  135552245  141897932    chrX      cX_b20\n",
       "973  974   23   21  141897933  152168662    chrX      cX_b21\n",
       "974  975   23   22  152168663  153788223    chrX      cX_b22\n",
       "975  976   23   23  153788224  156040895    chrX      cX_b23\n",
       "976  977   24    0          1   57227415    chrY       cY_b0\n",
       "\n",
       "[977 rows x 7 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blk = pd.read_table(\"./resources/pvcf_blocks.txt\", sep = '\\t', names = ['ind', 'chr', 'blk', 'start_pos', 'end_pos'])\n",
    "df_blk['seqname'] = 'chr' + df_blk['chr'].map(str)\n",
    "df_blk.loc[df_blk['seqname'] == 'chr23', 'seqname'] = 'chrX'\n",
    "df_blk.loc[df_blk['seqname'] == 'chr24', 'seqname'] = 'chrY'\n",
    "df_blk['chr_blk_str'] = df_blk['seqname'].str.replace('chr', 'c') + '_b' + df_blk['blk'].map(str)\n",
    "df_blk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16fc2f8f-3742-4c4e-b01e-05d35069d52d",
   "metadata": {},
   "source": [
    "### Load MANE transcript coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7c71254d-0da0-4458-92dd-3fcdf52e9b7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_flank_start</th>\n",
       "      <th>exon_flank_end</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>174707</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160664166</td>\n",
       "      <td>160664275</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160664161</td>\n",
       "      <td>160664280</td>\n",
       "      <td>chr6:160664161-160664280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174710</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160650338</td>\n",
       "      <td>160650497</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160650333</td>\n",
       "      <td>160650502</td>\n",
       "      <td>chr6:160650333-160650502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174712</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160646214</td>\n",
       "      <td>160646395</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160646209</td>\n",
       "      <td>160646400</td>\n",
       "      <td>chr6:160646209-160646400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174714</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160644847</td>\n",
       "      <td>160645006</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160644842</td>\n",
       "      <td>160645011</td>\n",
       "      <td>chr6:160644842-160645011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174716</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160640667</td>\n",
       "      <td>160640848</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160640662</td>\n",
       "      <td>160640853</td>\n",
       "      <td>chr6:160640662-160640853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174718</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160639300</td>\n",
       "      <td>160639459</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160639295</td>\n",
       "      <td>160639464</td>\n",
       "      <td>chr6:160639295-160639464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174720</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160635123</td>\n",
       "      <td>160635304</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160635118</td>\n",
       "      <td>160635309</td>\n",
       "      <td>chr6:160635118-160635309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174722</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160633753</td>\n",
       "      <td>160633912</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160633748</td>\n",
       "      <td>160633917</td>\n",
       "      <td>chr6:160633748-160633917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174724</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160629574</td>\n",
       "      <td>160629755</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160629569</td>\n",
       "      <td>160629760</td>\n",
       "      <td>chr6:160629569-160629760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174726</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160628207</td>\n",
       "      <td>160628366</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160628202</td>\n",
       "      <td>160628371</td>\n",
       "      <td>chr6:160628202-160628371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174728</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160624030</td>\n",
       "      <td>160624211</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160624025</td>\n",
       "      <td>160624216</td>\n",
       "      <td>chr6:160624025-160624216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174730</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160622663</td>\n",
       "      <td>160622822</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160622658</td>\n",
       "      <td>160622827</td>\n",
       "      <td>chr6:160622658-160622827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174732</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160618484</td>\n",
       "      <td>160618665</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160618479</td>\n",
       "      <td>160618670</td>\n",
       "      <td>chr6:160618479-160618670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174734</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160617117</td>\n",
       "      <td>160617276</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160617112</td>\n",
       "      <td>160617281</td>\n",
       "      <td>chr6:160617112-160617281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174736</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160612932</td>\n",
       "      <td>160613113</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160612927</td>\n",
       "      <td>160613118</td>\n",
       "      <td>chr6:160612927-160613118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174738</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160611562</td>\n",
       "      <td>160611721</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160611557</td>\n",
       "      <td>160611726</td>\n",
       "      <td>chr6:160611557-160611726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174740</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160606477</td>\n",
       "      <td>160606658</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160606472</td>\n",
       "      <td>160606663</td>\n",
       "      <td>chr6:160606472-160606663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174742</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160605046</td>\n",
       "      <td>160605205</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160605041</td>\n",
       "      <td>160605210</td>\n",
       "      <td>chr6:160605041-160605210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174744</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160600917</td>\n",
       "      <td>160601098</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160600912</td>\n",
       "      <td>160601103</td>\n",
       "      <td>chr6:160600912-160601103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174746</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160599500</td>\n",
       "      <td>160599659</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160599495</td>\n",
       "      <td>160599664</td>\n",
       "      <td>chr6:160599495-160599664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174748</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160595354</td>\n",
       "      <td>160595535</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160595349</td>\n",
       "      <td>160595540</td>\n",
       "      <td>chr6:160595349-160595540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174750</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160593958</td>\n",
       "      <td>160594117</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160593953</td>\n",
       "      <td>160594122</td>\n",
       "      <td>chr6:160593953-160594122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174752</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160590944</td>\n",
       "      <td>160591101</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160590939</td>\n",
       "      <td>160591106</td>\n",
       "      <td>chr6:160590939-160591106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174754</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160589553</td>\n",
       "      <td>160589712</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160589548</td>\n",
       "      <td>160589717</td>\n",
       "      <td>chr6:160589548-160589717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174756</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160586449</td>\n",
       "      <td>160586630</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160586444</td>\n",
       "      <td>160586635</td>\n",
       "      <td>chr6:160586444-160586635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174758</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160585046</td>\n",
       "      <td>160585205</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160585041</td>\n",
       "      <td>160585210</td>\n",
       "      <td>chr6:160585041-160585210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174760</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160578523</td>\n",
       "      <td>160578704</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160578518</td>\n",
       "      <td>160578709</td>\n",
       "      <td>chr6:160578518-160578709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174762</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160577136</td>\n",
       "      <td>160577295</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160577131</td>\n",
       "      <td>160577300</td>\n",
       "      <td>chr6:160577131-160577300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174764</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160557390</td>\n",
       "      <td>160557571</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160557385</td>\n",
       "      <td>160557576</td>\n",
       "      <td>chr6:160557385-160557576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174766</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160556025</td>\n",
       "      <td>160556184</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160556020</td>\n",
       "      <td>160556189</td>\n",
       "      <td>chr6:160556020-160556189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174768</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160548478</td>\n",
       "      <td>160548659</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160548473</td>\n",
       "      <td>160548664</td>\n",
       "      <td>chr6:160548473-160548664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174770</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160547789</td>\n",
       "      <td>160547937</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160547784</td>\n",
       "      <td>160547942</td>\n",
       "      <td>chr6:160547784-160547942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174772</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160545440</td>\n",
       "      <td>160545533</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160545435</td>\n",
       "      <td>160545538</td>\n",
       "      <td>chr6:160545435-160545538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174774</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160542688</td>\n",
       "      <td>160542808</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160542683</td>\n",
       "      <td>160542813</td>\n",
       "      <td>chr6:160542683-160542813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174776</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160541107</td>\n",
       "      <td>160541181</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160541102</td>\n",
       "      <td>160541186</td>\n",
       "      <td>chr6:160541102-160541186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174778</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160540043</td>\n",
       "      <td>160540183</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160540038</td>\n",
       "      <td>160540188</td>\n",
       "      <td>chr6:160540038-160540188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174780</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160537855</td>\n",
       "      <td>160537961</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160537850</td>\n",
       "      <td>160537966</td>\n",
       "      <td>chr6:160537850-160537966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174782</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160532531</td>\n",
       "      <td>160532649</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160532526</td>\n",
       "      <td>160532654</td>\n",
       "      <td>chr6:160532526-160532654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174784</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160531482</td>\n",
       "      <td>160531890</td>\n",
       "      <td>LPA</td>\n",
       "      <td>160531477</td>\n",
       "      <td>160531895</td>\n",
       "      <td>chr6:160531477-160531895</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname      start        end gene_name  exon_flank_start   \n",
       "174707    chr6  160664166  160664275       LPA         160664161  \\\n",
       "174710    chr6  160650338  160650497       LPA         160650333   \n",
       "174712    chr6  160646214  160646395       LPA         160646209   \n",
       "174714    chr6  160644847  160645006       LPA         160644842   \n",
       "174716    chr6  160640667  160640848       LPA         160640662   \n",
       "174718    chr6  160639300  160639459       LPA         160639295   \n",
       "174720    chr6  160635123  160635304       LPA         160635118   \n",
       "174722    chr6  160633753  160633912       LPA         160633748   \n",
       "174724    chr6  160629574  160629755       LPA         160629569   \n",
       "174726    chr6  160628207  160628366       LPA         160628202   \n",
       "174728    chr6  160624030  160624211       LPA         160624025   \n",
       "174730    chr6  160622663  160622822       LPA         160622658   \n",
       "174732    chr6  160618484  160618665       LPA         160618479   \n",
       "174734    chr6  160617117  160617276       LPA         160617112   \n",
       "174736    chr6  160612932  160613113       LPA         160612927   \n",
       "174738    chr6  160611562  160611721       LPA         160611557   \n",
       "174740    chr6  160606477  160606658       LPA         160606472   \n",
       "174742    chr6  160605046  160605205       LPA         160605041   \n",
       "174744    chr6  160600917  160601098       LPA         160600912   \n",
       "174746    chr6  160599500  160599659       LPA         160599495   \n",
       "174748    chr6  160595354  160595535       LPA         160595349   \n",
       "174750    chr6  160593958  160594117       LPA         160593953   \n",
       "174752    chr6  160590944  160591101       LPA         160590939   \n",
       "174754    chr6  160589553  160589712       LPA         160589548   \n",
       "174756    chr6  160586449  160586630       LPA         160586444   \n",
       "174758    chr6  160585046  160585205       LPA         160585041   \n",
       "174760    chr6  160578523  160578704       LPA         160578518   \n",
       "174762    chr6  160577136  160577295       LPA         160577131   \n",
       "174764    chr6  160557390  160557571       LPA         160557385   \n",
       "174766    chr6  160556025  160556184       LPA         160556020   \n",
       "174768    chr6  160548478  160548659       LPA         160548473   \n",
       "174770    chr6  160547789  160547937       LPA         160547784   \n",
       "174772    chr6  160545440  160545533       LPA         160545435   \n",
       "174774    chr6  160542688  160542808       LPA         160542683   \n",
       "174776    chr6  160541107  160541181       LPA         160541102   \n",
       "174778    chr6  160540043  160540183       LPA         160540038   \n",
       "174780    chr6  160537855  160537961       LPA         160537850   \n",
       "174782    chr6  160532531  160532649       LPA         160532526   \n",
       "174784    chr6  160531482  160531890       LPA         160531477   \n",
       "\n",
       "        exon_flank_end                    region  \n",
       "174707       160664280  chr6:160664161-160664280  \n",
       "174710       160650502  chr6:160650333-160650502  \n",
       "174712       160646400  chr6:160646209-160646400  \n",
       "174714       160645011  chr6:160644842-160645011  \n",
       "174716       160640853  chr6:160640662-160640853  \n",
       "174718       160639464  chr6:160639295-160639464  \n",
       "174720       160635309  chr6:160635118-160635309  \n",
       "174722       160633917  chr6:160633748-160633917  \n",
       "174724       160629760  chr6:160629569-160629760  \n",
       "174726       160628371  chr6:160628202-160628371  \n",
       "174728       160624216  chr6:160624025-160624216  \n",
       "174730       160622827  chr6:160622658-160622827  \n",
       "174732       160618670  chr6:160618479-160618670  \n",
       "174734       160617281  chr6:160617112-160617281  \n",
       "174736       160613118  chr6:160612927-160613118  \n",
       "174738       160611726  chr6:160611557-160611726  \n",
       "174740       160606663  chr6:160606472-160606663  \n",
       "174742       160605210  chr6:160605041-160605210  \n",
       "174744       160601103  chr6:160600912-160601103  \n",
       "174746       160599664  chr6:160599495-160599664  \n",
       "174748       160595540  chr6:160595349-160595540  \n",
       "174750       160594122  chr6:160593953-160594122  \n",
       "174752       160591106  chr6:160590939-160591106  \n",
       "174754       160589717  chr6:160589548-160589717  \n",
       "174756       160586635  chr6:160586444-160586635  \n",
       "174758       160585210  chr6:160585041-160585210  \n",
       "174760       160578709  chr6:160578518-160578709  \n",
       "174762       160577300  chr6:160577131-160577300  \n",
       "174764       160557576  chr6:160557385-160557576  \n",
       "174766       160556189  chr6:160556020-160556189  \n",
       "174768       160548664  chr6:160548473-160548664  \n",
       "174770       160547942  chr6:160547784-160547942  \n",
       "174772       160545538  chr6:160545435-160545538  \n",
       "174774       160542813  chr6:160542683-160542813  \n",
       "174776       160541186  chr6:160541102-160541186  \n",
       "174778       160540188  chr6:160540038-160540188  \n",
       "174780       160537966  chr6:160537850-160537966  \n",
       "174782       160532654  chr6:160532526-160532654  \n",
       "174784       160531895  chr6:160531477-160531895  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz\")\n",
    "df = df.loc[(df['feature'] == 'exon') & (df['gene_name'].isin(genes))]\n",
    "df = df[['seqname', 'start', 'end', 'gene_name']]\n",
    "df['exon_flank_start'] = df['start'] - exon_flank_nt\n",
    "df['exon_flank_end'] = df['end'] + exon_flank_nt\n",
    "df['region'] = ((df['seqname'] + ':').str.cat(df['exon_flank_start'].astype(str)) + '-').str.cat(df['exon_flank_end'].astype(str))\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e66005-ff41-4b75-a443-a2ed9df136bd",
   "metadata": {},
   "source": [
    "### Run Swiss-army-knife on DNAnexus\n",
    "- get region info for each gene from block file\n",
    "- bcftools command for step 2 finished\n",
    "- bcftools command for step 3 & 4 TBD\n",
    "- list genes not included in MANE set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "14ba3303-3410-4c56-a1f4-c6201572c1be",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GYz2zx0JqBj903ZYYVf98YXb\n",
      "\u001b[0mGenes not found in MANE database:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "genes_not_found = []\n",
    "genes_found = []\n",
    "known_large_genes = [\"DSP\", \"TSC2\", \"TTN\", \"NCOA3\"]\n",
    "\n",
    "for gene in genes:\n",
    "    df_gene = df.loc[df['gene_name'] == gene]\n",
    "    if df_gene.shape[0] > 0:\n",
    "        genes_found.append(gene)\n",
    "        \n",
    "        vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "        vcf_str = ' '.join(vcf_files)\n",
    "        region_str = ','.join(df_gene['region'].to_list())\n",
    "        \n",
    "        if ((len(vcf_files) > 1) or (gene in known_large_genes)):\n",
    "            mem_level = \"mem2_ssd1_v2_x16\" # dynamically change memory level when submitting jobs on DNAnexus\n",
    "        else:\n",
    "            mem_level = \"mem1_ssd1_v2_x4\"\n",
    "        \n",
    "        # filtering variants\n",
    "        bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "        bcftools_cmd2 = \"bcftools view -Ou --max-alleles 5 -T ^\" + diff_bed\n",
    "        bcftools_cmd3 = \"bcftools +fill-tags -Ou -- -t all\"\n",
    "        bcftools_cmd4 = \"bcftools norm -Ou -m - -f \" + ref_genome \n",
    "        bcftools_cmd5 = \"bcftools view -Ov -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \"_variants.vcf.gz\"\n",
    "        \n",
    "        # extract carriers\n",
    "        bcftools_cmd6 = \"mkdir -p carrier_out\"\n",
    "        bcftools_cmd7 = \"bcftools query -i 'GT=\\\\\\\"RA\\\\\\\"|GT=\\\\\\\"AR\\\\\\\"|GT=\\\\\\\"AA\\\\\\\"' -f '%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n' \" + gene + \"_variants.vcf.gz > carrier_out/\" + gene + \".ssv\"\n",
    "        \n",
    "        # parsing command\n",
    "        bcftools_command_a = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "        bcftools_command_b = \" && \".join([bcftools_cmd6, bcftools_cmd7])\n",
    "        bcftools_command = bcftools_command_a + \" && \" + bcftools_command_b\n",
    "        \n",
    "        # parsing input\n",
    "        dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "        \n",
    "        # final dx command\n",
    "        dx_command = 'dx run swiss-army-knife --instance-type ' + mem_level + ' -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "        !{dx_command}\n",
    "    else:\n",
    "        genes_not_found.append(gene)\n",
    "            \n",
    "print('Genes not found in MANE database:')\n",
    "print(genes_not_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "adf05525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dx run swiss-army-knife --instance-type mem2_ssd1_v2_x16 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr7:6008992-6009054,chr7:6005887-6006036,chr7:6003967-6004063,chr7:6003685-6003797,chr7:6002448-6002641,chr7:5999103-5999280,chr7:5997321-5997428,chr7:5995529-5995638,chr7:5991968-5992062,chr7:5989795-5989960,chr7:5986754-5987625,chr7:5982819-5982996,chr7:5978591-5978701,chr7:5977583-5977762,chr7:5970920-5973547 ukb23157_c7_b4_v1.vcf.gz ukb23157_c7_b3_v1.vcf.gz | bcftools view -Ou --max-alleles 5 -T ^GRCh38_alldifficultregions.bed.gz | bcftools +fill-tags -Ou -- -t all | bcftools norm -Oz -m - -f GRCh38_reference_genome.fa> filtered.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/Tian_folder/Vineel_genes_test/ --tag \"Vineel_090823\" --property gene=PMS2'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dx_command_c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e454b2db-5ba6-446f-8c3c-31c164c7df92",
   "metadata": {},
   "source": [
    "### Below is for testing purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4df214f4-69b6-4396-a167-e7a0ff599e41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\n",
      "dx run swiss-army-knife --instance-type mem2_ssd1_v2_x4 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/temp_vcf_out/ --tag \"my_tag\" --property gene=APOB\n"
     ]
    }
   ],
   "source": [
    "gene = 'APOB'\n",
    "df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "vcf_str = ' '.join(vcf_files)\n",
    "region_str = ','.join(df_gene['region'].to_list())\n",
    "mem_level = len(vcf_files)\n",
    "\n",
    "bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "bcftools_cmd2 = \"bcftools view -Ou -T ^\" + diff_bed\n",
    "bcftools_cmd3 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "bcftools_cmd4 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "bcftools_cmd5 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "\n",
    "dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "dx_command = 'dx run swiss-army-knife --instance-type mem' + str(mem_level) + '_ssd1_v2_x4 -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "\n",
    "print(bcftools_command)\n",
    "print(dx_command)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4d4d5547-7ac4-4a6b-8409-130c10019c9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GPG9KY8JqBj5gK412JXv3f8G\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!{dx_command}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
