{"metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.5"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# Generating Patient-Variant Mappings\n\n##### Updated 02/20/2023\n##### Lara Brown\n\n#### Goal:\n\nGenerate a csv file detailing the most severe variant for each UK Biobank patient for a given gene.\n\n#### Required inputs:\n\n**Variant-patient mapping:** You will need a gene-level variant patient mapping ssv file with the following columns: `Chrom, Name (variant id), Pos, Ref, Alt, Carriers`. The list of carriers should be pipe-separated, though the methods used in this file can be modified to match other formats. \n\n**Annotated variant file:** You will also need a csv file of VEP-annotated variants for the same gene. This file can be generated on the BWH genetics cluster by modifying and running the wrapper script at `/net/ukbb/500k_processing_lara/parse_vcf_parallel.sh`. This script yields a .csv file that can then be uploaded to the RAP and used for the methods in this notebook. The two main components of this script are:\n\n1. Running VEP using the following command:\n\n```\nvep --cache --force_overwrite --offline --dir /net/data/vep --assembly GRCh38 --plugin dbNSFP,/net/ukbb/data/dbnsfp_content/dbNSFP4.0a.gz,ALL --plugin LoF,loftee_path:/net/data/vep/loftee-grch38,human_ancestor_fa:/net/data/vep/loftee-grch38/human_ancestor.fa.gz --vcf --max_af --af_gnomade --af_gnomadg --no_check_variants_order --canonical --no_stats -I {gene_input.vcf.gz} -o {output_vep.vcf}\n```\n\n2. Parsing the VEP output by running `/net/ukbb/500k_processing_lara/vep_parser_af_lof.py`. Note that this parser requires that the VEP annotations include (a) max allele frequencies, both the default from dbNSFP (`allele_frequency`), as well as `gnomADe_MAX_AF` and `gnomADg_MAX_AF`, and (b) LoF confidence scores, obtained from the `loftee` VEP plugin. \n\n#### Generated outputs:\n\nThe methods in this notebook can be used to generate a gene-level csv file containing the following columns:\n\n| Column name | Description | Data type |\n| :- | :- | :-: |\n| eid | ID of carrier in the UK Biobank | object\n| variant_id/most_severe_variant | ID of the variant described in the remainder of the row. Format Chrom-Pos-Ref-Alt | object\n| Synonymous | 1 if variant is synonymous; 0 otherwise| int64\n| Missense | 1 if variant is missense; 0 otherwise| int64\n| Deleterious | 1 if variant is one of: `stop_gained, start_lost, splice_acceptor_variant, splice_donor_variant, frameshift_variant`; 0 otherwise| int64\n| coding_position | relative position of base pair in coding sequence | float64\n| AA_name | for non-synonymous variants, lists the amino acid substitution as RefLocationAlt | object\n\n#### Non-interactive option:\n\n**The code in this notebook is not actively maintained. Please use this option for the most updated version!** The outputs in this notebook can (and should!!) also be generated by running the script stored on RAP at `patient_variant_extracts/scripts/var-pt-map.py`. To do so, you will need to create a Docker image based on Python 3.6 that properly installs Pandas and Numpy. Then, run the following commands on a local terminal where Python SDK/dx-toolkit has been properly configured:\n\n```\ncmd=\"python3 var-pt-map.py LDLR -v blend_ukb/resources/parsed_bc_genes -s -e VEST4_score,CADD,REVEL_score,ukb_af\"\ndx run swiss-army-knife -iin=\"file-GQVXx8jJqBj0Vg55zP03kXj5\" -icmd=\"$cmd\" -iimage_file=\"patient_variant_extracts/scripts/pandas_numpy.tar.gz\" --destination \"patient_variant_extracts/bc_genes\" -y\n```\n\nThe file ID specified as the Swiss Army Knife input corresponds to the aforementioned `var-pt-map.py` script. Replace LDLR with your gene of choice, or run in a for-loop over multiple genes. The Python script is set up to accept the following flags:\n\n| Flag | Full name | Description | Data type |\n| :- | :- | :- | :-: |\n| gene | n/a | Required; symbol of gene to access | str \n| -v | --var_path | Name of directory in which annotated variant file is stored. Assumes file name is {gene}_parsed.csv | str\n| -c | --car_path | Name of directory in which variant-patient `.ssv` mapping is stored. Assumes file name is {gene}.ssv | str\n| -s | --most_severe | Includes only most severe variant per patient when specified | store_true\n| -e | --extra_cols | Comma delimited list of additional VEP columns to include in output. Use to include, eg, computational scores like VEST4 or CADD9 in the output file. ukb_af can be included if it was retained at initial generation of var-pt SSVs. | comma separated str (converted to list) |\n", "metadata": {}}, {"cell_type": "code", "source": "import pandas as pd\nimport numpy as np\nfrom pandas.api.types import CategoricalDtype\nfrom random import shuffle", "metadata": {"trusted": true}, "execution_count": 1, "outputs": []}, {"cell_type": "markdown", "source": "### Packaging in functions", "metadata": {}}, {"cell_type": "code", "source": "def get_variants(gene, dir_path=\"selected_genes/parsed_selected_genes_filtered_variants_5nt_flank\"):\n    \"\"\"retrieves annotated and parsed variant file\"\"\"\n    \n    path = f'/mnt/project/{dir_path}/{gene}_parsed.csv'\n    \n    return(pd.read_csv(path))\n    ", "metadata": {"trusted": true}, "execution_count": 2, "outputs": []}, {"cell_type": "code", "source": "def get_carriers(gene, dir_path=\"selected_genes/selected_genes_var_patient_5nt_flank\"):\n    \"\"\"retrieve variant-patient mapping file\"\"\"\n    \n    path = f'/mnt/project/{dir_path}/{gene}.ssv'\n    carriers = pd.read_csv(path,\n                       sep=\" \",\n                       names=[\"Chrom\", \"na\", \"Pos\", \"Ref\", \"Alt\", \"Carriers\"],\n                       usecols=lambda x: x != 'na')\n    carriers.loc[:, \"Carriers\"] = carriers[\"Carriers\"].str.strip(\"|\")\n    carriers.loc[:, \"Carriers\"]= carriers[\"Carriers\"].str.split(\"|\", expand = False)\n    carriers = carriers.explode(\"Carriers\").rename(columns={\"Carriers\": \"Carrier\"})\n    carriers = carriers[carriers['Carrier'].apply(lambda x: not x.startswith('W'))] # remove withdrawn\n    carriers[\"Chrom\"] = carriers[\"Chrom\"].str.extract('(\\d+)').astype(\"int64\")\n    \n    return(carriers)", "metadata": {"trusted": true}, "execution_count": 3, "outputs": []}, {"cell_type": "code", "source": "def get_joined(gene, **paths):\n    \"\"\"retrieves both variant and mapping files and joins them together by variant\"\"\"\n    \n    variants = get_variants(gene, paths[\"var_path\"]) if \"var_path\" in paths else get_variants(gene)\n    carriers = get_carriers(gene, paths[\"car_path\"]) if \"car_path\" in paths else get_carriers(gene)\n    ann_carriers = carriers.join(variants.set_index([\"Chrom\", \"Pos\", \"Ref\", \"Alt\"]), on=[\"Chrom\", \"Pos\", \"Ref\", \"Alt\"])\n    \n    return(ann_carriers)", "metadata": {"trusted": true}, "execution_count": 4, "outputs": []}, {"cell_type": "markdown", "source": "Note that `get_select_carriers` below uses the Deleterious category as defined in the VEP parser, which includes splice acceptor/donor variants, frameshift, stop gained, and start lost. The list in the first version of this function excludes start lost variants, so the total number of records will differ depending on which version is used. Also, the Deleterious category contains variants annotated with multiple consequences, at least one of which is deleterious, whereas the list from the other version of this function requires a perfect string match and thus would not count any variant with multiple consequences as deleterious. This latter criterion, for example, adds an extra 1,008 deleterious variant/carrier pairs in LDLR when including doubly-annotated consequences.", "metadata": {}}, {"cell_type": "code", "source": "def get_select_carriers(ann_carriers):\n    \"\"\"\n    accepts a joined carrier-variant df.\n    filters by consequence and creates a string to represent amino acid substitutions when present\n    \"\"\"\n    ann_carriers = ann_carriers.loc[(ann_carriers[\"consequence\"].isin([\"Missense\", \"Synonymous\"])) |\n                                   ((ann_carriers[\"consequence\"] == \"Deleterious\") & (ann_carriers[\"LoF_confidence\"] == \"HC\"))]\n    ann_carriers._is_copy = False # gets rid of SettingWithCopyWarning\n    ann_carriers.loc[ann_carriers.consequence == \"Deleterious\", \"VEST4_score\"] = 1 # set vest4 to max for del variants\n\n    if ann_carriers.loc[:, \"AA_pos\"].dtypes != \"object\":\n        ann_carriers.loc[:,\"AA_pos\"] = ann_carriers[\"AA_pos\"].astype('Int64').astype('str')\n    ann_carriers.loc[ann_carriers.AA_alt == \"Stop\", \"AA_alt\"] = \"*\"\n    ann_carriers.loc[ann_carriers.AA_ref == \"Stop\", \"AA_ref\"] = \"*\"\n    ann_carriers.loc[:, \"AA_name\"] = ann_carriers[\"AA_ref\"] + ann_carriers[\"AA_pos\"] + ann_carriers[\"AA_alt\"]\n    ann_carriers.loc[ann_carriers.consequence == \"Synonymous\", \"AA_name\"] = np.NaN\n\n    return(ann_carriers)\n", "metadata": {"trusted": true}, "execution_count": 5, "outputs": []}, {"cell_type": "code", "source": "def prep_for_pivot(ann_carriers):\n    \"\"\"accepts a joined and filtered patient-variant df and fills in missing values to allow pandas pivot\"\"\"\n    \n    ann_carriers.loc[ann_carriers[\"CADD\"].isnull(), \"CADD\"] = \"dummy\"\n    ann_carriers.loc[ann_carriers[\"VEST4_score\"].isnull(), \"VEST4_score\"] = \"dummy\"\n    ann_carriers.loc[ann_carriers[\"coding_position\"].isnull(), \"coding_position\"] = \"dummy\"\n    ann_carriers.loc[ann_carriers[\"AA_name\"].isnull(), \"AA_name\"] = \"dummy\"\n    ann_carriers.loc[:, \"val\"] = 1\n\n    return(ann_carriers)", "metadata": {"trusted": true}, "execution_count": 6, "outputs": []}, {"cell_type": "code", "source": "def get_most_severe(ann_carriers):\n    \"\"\"\n    accepts a filtered patient-variant df.\n    selects and retains only the most severe variant per patient. returns this df.\n    \"\"\"\n    \n    # order variant consequence by general severity (deleterious > missense > synonymous)\n    consequence_cats = CategoricalDtype(categories=[\"Synonymous\", \"Missense\", \"Deleterious\"], ordered=True)\n    ann_carriers.loc[:, \"consequence\"] = ann_carriers[\"consequence\"].astype(consequence_cats)\n    \n    # choose maximum AF for each variant, where possible\n    ann_carriers.loc[:, \"max_AF\"] = ann_carriers.loc[:, [\"allele_frequency\", \"gnomADe_MAX_AF\", \"gnomADg_MAX_AF\"]].max(axis=1, skipna=True)\n    \n    # order deleterious variance confidence levels (high > low)\n    del_confidence_cats = CategoricalDtype(categories=[\"LC\", \"HC\"], ordered=True)\n    ann_carriers.loc[:, \"LoF_confidence\"] = ann_carriers[\"LoF_confidence\"].astype(del_confidence_cats)\n \n    # create a random tie breaker index\n    shuffled_index = list(range(0,len(ann_carriers)))\n    shuffle(shuffled_index)\n    ann_carriers.loc[:, \"tie_breaker\"] = shuffled_index\n    \n    # group by eid, sort by consequence > CADD > confidence (for del. vars) > AF > tiebreaker\n    # select top variant for each group\n    ann_carriers_grouped = ann_carriers.groupby(\"Carrier\", group_keys=False).apply(pd.DataFrame.sort_values, [\"consequence\", \"CADD\", \"LoF_confidence\", \"max_AF\", \"tie_breaker\"], ascending=[False, False, False, True, True])\n    ann_carriers_grouped = ann_carriers_grouped.drop_duplicates(\"Carrier\", keep=\"first\")\n    \n    # important!! return categorical variables to normal objects so they don't clog pivot memory\n    ann_carriers_grouped.loc[:, \"consequence\"] = ann_carriers_grouped[\"consequence\"].astype(object)\n    ann_carriers_grouped.loc[:, \"LoF_confidence\"] = ann_carriers_grouped[\"LoF_confidence\"].astype(object)\n    \n    return(ann_carriers_grouped)", "metadata": {"trusted": true}, "execution_count": 7, "outputs": []}, {"cell_type": "code", "source": "def reshape_consequence(ann_carriers, rename_id=False):\n    \"\"\"\n    accepts filtered and prepped patient-variant df\n    returns a reshaped df that separates consequences into unique columns\n    \"\"\"\n    \n    ann_carriers = ann_carriers.pivot_table(index=[\"Carrier\", \"Name\", \"coding_position\", \"CADD\", \"VEST4_score\", \"AA_name\"], values=\"val\", columns=\"consequence\", fill_value=0)\n    ann_carriers = ann_carriers.reset_index().replace('dummy',np.nan)\n\n    new_names={\"Carrier\": \"eid\", \"Name\": \"variant_id\"}\n    \n    # this variable is set to true whenever get_most_severe was run before this function\n    if rename_id:\n        new_names[\"Name\"] = \"most_severe_variant\"\n\n    selected_cols=[\"Carrier\", \"Name\", \"Synonymous\", \"Missense\", \"Deleterious\", \"coding_position\", \"AA_name\"]\n    \n    ann_carriers = ann_carriers.reindex(columns=selected_cols)\n    ann_carriers = ann_carriers.rename(columns=new_names)\n    ann_carriers.columns.name = None\n    \n    return(ann_carriers)", "metadata": {"trusted": true}, "execution_count": 8, "outputs": []}, {"cell_type": "code", "source": "def patient_var_mappings(gene, most_severe=False, **paths):\n    \"\"\"\n    accepts gene name and some combination of var_path and car_path directories.\n    returns a pandas df with one row per patient detailing their most severe variant\n    \"\"\"\n    annotated_carriers = get_joined(gene, **paths)\n    annotated_carriers = get_select_carriers(annotated_carriers)\n    annotated_carriers = prep_for_pivot(annotated_carriers)\n    \n    if most_severe:\n        annotated_carriers = get_most_severe(annotated_carriers)\n        \n    return(reshape_consequence(annotated_carriers, rename_id=most_severe))", "metadata": {"trusted": true}, "execution_count": 9, "outputs": []}, {"cell_type": "markdown", "source": "### Obtaining extracts!", "metadata": {}}, {"cell_type": "code", "source": "genes = [\"TP53\", \"BRCA1\", \"BRCA2\", \"LDLR\"]\nfor g in genes:\n    file_name=f'{g}.csv'\n    patient_var_mappings(g, var_path=\"parsed_bc_genes\").to_csv(file_name, index=False)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# adapt this code to test the sorting process for patients with multiple variants\n# test = ldlr_carriers_grouped[[\"Carrier\", \"Name\", \"vep_consequence\", \"VEST4_score\", \"CADD\", \"LoF_confidence\", \"max_AF\", \"tie_breaker\"]]\n# count = ldlr_carriers.groupby(\"Carrier\").size().where(lambda x : x>2).dropna()\n# c = ldlr_carriers.loc[ldlr_carriers['LoF_confidence'].notna()][\"Carrier\"].unique()\n\n# multiple_var_ids = count.index.values.tolist()\n# test[test[\"Carrier\"].isin(multiple_var_ids)].head(25)", "metadata": {"trusted": true}, "execution_count": 46, "outputs": [{"execution_count": 46, "output_type": "execute_result", "data": {"text/plain": "     Carrier             Name     vep_consequence  VEST4_score      CADD  \\\n243  1237033  19-11110681-G-A    missense_variant        0.549  3.861107   \n186  1237033  19-11107402-C-T  synonymous_variant          NaN       NaN   \n194  1237033  19-11107432-C-T  synonymous_variant          NaN       NaN   \n186  1367116  19-11107402-C-T  synonymous_variant          NaN       NaN   \n194  1367116  19-11107432-C-T  synonymous_variant          NaN       NaN   \n7    1367116  19-11100245-C-T  synonymous_variant          NaN       NaN   \n265  1451411  19-11110735-G-A    missense_variant        0.187  1.391810   \n175  1451411  19-11106660-A-T    missense_variant        0.284  1.081236   \n650  1451411  19-11120455-G-A  synonymous_variant          NaN       NaN   \n265  1623518  19-11110735-G-A    missense_variant        0.187  1.391810   \n175  1623518  19-11106660-A-T    missense_variant        0.284  1.081236   \n650  1623518  19-11120455-G-A  synonymous_variant          NaN       NaN   \n162  1997600  19-11106627-C-T    missense_variant        0.729  3.211820   \n186  1997600  19-11107402-C-T  synonymous_variant          NaN       NaN   \n194  1997600  19-11107432-C-T  synonymous_variant          NaN       NaN   \n70   2456674  19-11102758-C-T  synonymous_variant          NaN       NaN   \n186  2456674  19-11107402-C-T  synonymous_variant          NaN       NaN   \n194  2456674  19-11107432-C-T  synonymous_variant          NaN       NaN   \n70   2813607  19-11102758-C-T  synonymous_variant          NaN       NaN   \n186  2813607  19-11107402-C-T  synonymous_variant          NaN       NaN   \n194  2813607  19-11107432-C-T  synonymous_variant          NaN       NaN   \n70   2977321  19-11102758-C-T  synonymous_variant          NaN       NaN   \n186  2977321  19-11107402-C-T  synonymous_variant          NaN       NaN   \n194  2977321  19-11107432-C-T  synonymous_variant          NaN       NaN   \n584  3249102  19-11120122-G-A    missense_variant        0.676  3.689852   \n\n    LoF_confidence    max_AF  tie_breaker  \n243            NaN  0.014703         9912  \n186            NaN  0.003629         9745  \n194            NaN  0.004122         7334  \n186            NaN  0.003629        11015  \n194            NaN  0.004122        11122  \n7              NaN  0.031150         1198  \n265            NaN  0.006591         3412  \n175            NaN       NaN         8550  \n650            NaN  0.000131         1457  \n265            NaN  0.006591         9171  \n175            NaN       NaN        10055  \n650            NaN  0.000131         1480  \n162            NaN  0.001230         2613  \n186            NaN  0.003629         5956  \n194            NaN  0.004122         4532  \n70             NaN  0.000184         2288  \n186            NaN  0.003629         6906  \n194            NaN  0.004122         7170  \n70             NaN  0.000184          670  \n186            NaN  0.003629          247  \n194            NaN  0.004122         8089  \n70             NaN  0.000184          276  \n186            NaN  0.003629         1961  \n194            NaN  0.004122         9607  \n584            NaN  0.000615         6455  ", "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Carrier</th>\n      <th>Name</th>\n      <th>vep_consequence</th>\n      <th>VEST4_score</th>\n      <th>CADD</th>\n      <th>LoF_confidence</th>\n      <th>max_AF</th>\n      <th>tie_breaker</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>243</th>\n      <td>1237033</td>\n      <td>19-11110681-G-A</td>\n      <td>missense_variant</td>\n      <td>0.549</td>\n      <td>3.861107</td>\n      <td>NaN</td>\n      <td>0.014703</td>\n      <td>9912</td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>1237033</td>\n      <td>19-11107402-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.003629</td>\n      <td>9745</td>\n    </tr>\n    <tr>\n      <th>194</th>\n      <td>1237033</td>\n      <td>19-11107432-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.004122</td>\n      <td>7334</td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>1367116</td>\n      <td>19-11107402-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.003629</td>\n      <td>11015</td>\n    </tr>\n    <tr>\n      <th>194</th>\n      <td>1367116</td>\n      <td>19-11107432-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.004122</td>\n      <td>11122</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>1367116</td>\n      <td>19-11100245-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.031150</td>\n      <td>1198</td>\n    </tr>\n    <tr>\n      <th>265</th>\n      <td>1451411</td>\n      <td>19-11110735-G-A</td>\n      <td>missense_variant</td>\n      <td>0.187</td>\n      <td>1.391810</td>\n      <td>NaN</td>\n      <td>0.006591</td>\n      <td>3412</td>\n    </tr>\n    <tr>\n      <th>175</th>\n      <td>1451411</td>\n      <td>19-11106660-A-T</td>\n      <td>missense_variant</td>\n      <td>0.284</td>\n      <td>1.081236</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>8550</td>\n    </tr>\n    <tr>\n      <th>650</th>\n      <td>1451411</td>\n      <td>19-11120455-G-A</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.000131</td>\n      <td>1457</td>\n    </tr>\n    <tr>\n      <th>265</th>\n      <td>1623518</td>\n      <td>19-11110735-G-A</td>\n      <td>missense_variant</td>\n      <td>0.187</td>\n      <td>1.391810</td>\n      <td>NaN</td>\n      <td>0.006591</td>\n      <td>9171</td>\n    </tr>\n    <tr>\n      <th>175</th>\n      <td>1623518</td>\n      <td>19-11106660-A-T</td>\n      <td>missense_variant</td>\n      <td>0.284</td>\n      <td>1.081236</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>10055</td>\n    </tr>\n    <tr>\n      <th>650</th>\n      <td>1623518</td>\n      <td>19-11120455-G-A</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.000131</td>\n      <td>1480</td>\n    </tr>\n    <tr>\n      <th>162</th>\n      <td>1997600</td>\n      <td>19-11106627-C-T</td>\n      <td>missense_variant</td>\n      <td>0.729</td>\n      <td>3.211820</td>\n      <td>NaN</td>\n      <td>0.001230</td>\n      <td>2613</td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>1997600</td>\n      <td>19-11107402-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.003629</td>\n      <td>5956</td>\n    </tr>\n    <tr>\n      <th>194</th>\n      <td>1997600</td>\n      <td>19-11107432-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.004122</td>\n      <td>4532</td>\n    </tr>\n    <tr>\n      <th>70</th>\n      <td>2456674</td>\n      <td>19-11102758-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.000184</td>\n      <td>2288</td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>2456674</td>\n      <td>19-11107402-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.003629</td>\n      <td>6906</td>\n    </tr>\n    <tr>\n      <th>194</th>\n      <td>2456674</td>\n      <td>19-11107432-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.004122</td>\n      <td>7170</td>\n    </tr>\n    <tr>\n      <th>70</th>\n      <td>2813607</td>\n      <td>19-11102758-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.000184</td>\n      <td>670</td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>2813607</td>\n      <td>19-11107402-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.003629</td>\n      <td>247</td>\n    </tr>\n    <tr>\n      <th>194</th>\n      <td>2813607</td>\n      <td>19-11107432-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.004122</td>\n      <td>8089</td>\n    </tr>\n    <tr>\n      <th>70</th>\n      <td>2977321</td>\n      <td>19-11102758-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.000184</td>\n      <td>276</td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>2977321</td>\n      <td>19-11107402-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.003629</td>\n      <td>1961</td>\n    </tr>\n    <tr>\n      <th>194</th>\n      <td>2977321</td>\n      <td>19-11107432-C-T</td>\n      <td>synonymous_variant</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0.004122</td>\n      <td>9607</td>\n    </tr>\n    <tr>\n      <th>584</th>\n      <td>3249102</td>\n      <td>19-11120122-G-A</td>\n      <td>missense_variant</td>\n      <td>0.676</td>\n      <td>3.689852</td>\n      <td>NaN</td>\n      <td>0.000615</td>\n      <td>6455</td>\n    </tr>\n  </tbody>\n</table>\n</div>"}, "metadata": {}}]}, {"cell_type": "code", "source": "patient_var_mappings(\"TP53\", var_path=\"cancer_func_score/parsed_bc_genes\", most_severe=True)", "metadata": {"trusted": true}, "execution_count": 13, "outputs": [{"execution_count": 13, "output_type": "execute_result", "data": {"text/plain": "          eid most_severe_variant  Synonymous  Missense  Deleterious  \\\n0     1001862      17-7669642-G-A           1         0            0   \n1     1002783      17-7670613-A-C           0         1            0   \n2     1004632      17-7676159-A-T           1         0            0   \n3     1006957      17-7676589-C-T           1         0            0   \n4     1009560      17-7670684-C-A           0         1            0   \n...       ...                 ...         ...       ...          ...   \n3817  6022267      17-7676047-C-G           0         1            0   \n3818  6022834      17-7676230-G-A           0         1            0   \n3819  6023144      17-7673751-C-T           0         1            0   \n3820  6023387      17-7676589-C-T           1         0            0   \n3821  6023750      17-7673717-T-C           1         0            0   \n\n      coding_position AA_name  \n0              1149.0     NaN  \n1              1096.0   S234A  \n2               210.0     NaN  \n3                 6.0     NaN  \n4              1025.0   R210L  \n...               ...     ...  \n3817            322.0   G108R  \n3818            139.0    P47S  \n3819            869.0   R290H  \n3820              6.0     NaN  \n3821            903.0     NaN  \n\n[3822 rows x 7 columns]", "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>eid</th>\n      <th>most_severe_variant</th>\n      <th>Synonymous</th>\n      <th>Missense</th>\n      <th>Deleterious</th>\n      <th>coding_position</th>\n      <th>AA_name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1001862</td>\n      <td>17-7669642-G-A</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1149.0</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1002783</td>\n      <td>17-7670613-A-C</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>1096.0</td>\n      <td>S234A</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>1004632</td>\n      <td>17-7676159-A-T</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>210.0</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>1006957</td>\n      <td>17-7676589-C-T</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6.0</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>1009560</td>\n      <td>17-7670684-C-A</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>1025.0</td>\n      <td>R210L</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>3817</th>\n      <td>6022267</td>\n      <td>17-7676047-C-G</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>322.0</td>\n      <td>G108R</td>\n    </tr>\n    <tr>\n      <th>3818</th>\n      <td>6022834</td>\n      <td>17-7676230-G-A</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>139.0</td>\n      <td>P47S</td>\n    </tr>\n    <tr>\n      <th>3819</th>\n      <td>6023144</td>\n      <td>17-7673751-C-T</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>869.0</td>\n      <td>R290H</td>\n    </tr>\n    <tr>\n      <th>3820</th>\n      <td>6023387</td>\n      <td>17-7676589-C-T</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6.0</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3821</th>\n      <td>6023750</td>\n      <td>17-7673717-T-C</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>903.0</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n<p>3822 rows \u00d7 7 columns</p>\n</div>"}, "metadata": {}}]}]}