{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a9a7db64-5615-462a-adc3-446da0767246",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "from subprocess import call\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8edf64c6-81e8-412c-968d-115baeaa3be2",
   "metadata": {},
   "source": [
    "### Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\n",
    "number_of_threads = 4 # number of threads used in bcftools output compression\n",
    "tag_str = 'Tian_011624' # DNAnexus job tag\n",
    "\n",
    "project_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n",
    "\n",
    "# input vcf path (end with '/')\n",
    "dx_vcf_path = project_path + \"Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/\"\n",
    "# output vcf path (end with '/')\n",
    "dx_vcf_out_path = project_path + \"Tian_folder/Tian_011624/\"\n",
    "# resource path (end with '/')\n",
    "dx_resource_path = project_path + \"GRCh38_resources/\"\n",
    "# difficult regions bed filename\n",
    "diff_bed = 'GRCh38_alldifficultregions.bed.gz'\n",
    "# reference genome filename\n",
    "ref_genome = 'GRCh38_reference_genome.fa' # index filename is inferred\n",
    "\n",
    "!dx mkdir -p {dx_vcf_out_path} # create gene.vcf output folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4000040-9b97-4085-a63e-b5aec384499f",
   "metadata": {},
   "source": [
    "### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2a8f096b-5b91-43aa-b546-4eaec039c30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_UKB_vcfs(df_gene, df_blk, diagnosis=False):\n",
    "    vcf_prefix = 'ukb23157_'\n",
    "    vcf_suffix = '_v1.vcf.gz'\n",
    "    \n",
    "    all_vcf_files = []\n",
    "    df_gene = df_gene.assign(vcf_files = '')\n",
    "    for index, row in df_gene.iterrows():\n",
    "        new_vcf_files = df_blk.loc[(df_blk['seqname'] == row['seqname']) & \n",
    "                                   ((df_blk['end_pos'] >= row['exon_flank_start']) & (df_blk['start_pos'] <= row['exon_flank_end'])),\n",
    "                                   'chr_blk_str'].tolist()\n",
    "        new_vcf_files = list(vcf_prefix + x + vcf_suffix for x in set(new_vcf_files))\n",
    "        df_gene.loc[index,'vcf_files'] = ','.join(new_vcf_files)\n",
    "        all_vcf_files = all_vcf_files + new_vcf_files\n",
    "    all_vcf_files = list(set(all_vcf_files))\n",
    "    \n",
    "    if diagnosis:\n",
    "        return(df_gene)\n",
    "    else:\n",
    "        return(all_vcf_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed62c94-9b34-42e2-90eb-67a527b04894",
   "metadata": {},
   "source": [
    "### List of gene symbols as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8fbe087c-8e0b-4b71-9f81-d9fbada151d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"TSR2\", \"IGSF23\", \"PREB\", \"TICRR\", \"PYGO2\", \"TRIQK\", \"DR1\", \"DGAT2\", \"MBOAT7\", \"PLIN2\"]\n",
    "\n",
    "# # to read from txt:\n",
    "# with open(\"resources/my_gene_file.txt\", \"r\") as gene_file:\n",
    "#     genes = gene_file.readlines()\n",
    "#     genes = [l.replace(\"\\n\", \"\") for l in genes]\n",
    "#     print(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "214b5b7e-a208-4c88-bf56-19dbcf0034bd",
   "metadata": {},
   "source": [
    "### Load pVCF block coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "62a230e4-d95d-4084-a914-01abecc3ff92",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>chr</th>\n",
       "      <th>blk</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>seqname</th>\n",
       "      <th>chr_blk_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1218130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1218131</td>\n",
       "      <td>1426969</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1426970</td>\n",
       "      <td>1758871</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1758872</td>\n",
       "      <td>2514221</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2514222</td>\n",
       "      <td>3782130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>973</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>135552245</td>\n",
       "      <td>141897932</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>974</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>141897933</td>\n",
       "      <td>152168662</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>975</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>152168663</td>\n",
       "      <td>153788223</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>976</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>153788224</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>977</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "      <td>cY_b0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>977 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ind  chr  blk  start_pos    end_pos seqname chr_blk_str\n",
       "0      1    1    0          1    1218130    chr1       c1_b0\n",
       "1      2    1    1    1218131    1426969    chr1       c1_b1\n",
       "2      3    1    2    1426970    1758871    chr1       c1_b2\n",
       "3      4    1    3    1758872    2514221    chr1       c1_b3\n",
       "4      5    1    4    2514222    3782130    chr1       c1_b4\n",
       "..   ...  ...  ...        ...        ...     ...         ...\n",
       "972  973   23   20  135552245  141897932    chrX      cX_b20\n",
       "973  974   23   21  141897933  152168662    chrX      cX_b21\n",
       "974  975   23   22  152168663  153788223    chrX      cX_b22\n",
       "975  976   23   23  153788224  156040895    chrX      cX_b23\n",
       "976  977   24    0          1   57227415    chrY       cY_b0\n",
       "\n",
       "[977 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blk = pd.read_table(\"./resources/pvcf_blocks.txt\", sep = '\\t', names = ['ind', 'chr', 'blk', 'start_pos', 'end_pos'])\n",
    "df_blk['seqname'] = 'chr' + df_blk['chr'].map(str)\n",
    "df_blk.loc[df_blk['seqname'] == 'chr23', 'seqname'] = 'chrX'\n",
    "df_blk.loc[df_blk['seqname'] == 'chr24', 'seqname'] = 'chrY'\n",
    "df_blk['chr_blk_str'] = df_blk['seqname'].str.replace('chr', 'c') + '_b' + df_blk['blk'].map(str)\n",
    "df_blk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16fc2f8f-3742-4c4e-b01e-05d35069d52d",
   "metadata": {},
   "source": [
    "### Load MANE transcript coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7c71254d-0da0-4458-92dd-3fcdf52e9b7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_flank_start</th>\n",
       "      <th>exon_flank_end</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22245</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93345907</td>\n",
       "      <td>93346865</td>\n",
       "      <td>DR1</td>\n",
       "      <td>93345902</td>\n",
       "      <td>93346870</td>\n",
       "      <td>chr1:93345902-93346870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22248</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93353908</td>\n",
       "      <td>93354071</td>\n",
       "      <td>DR1</td>\n",
       "      <td>93353903</td>\n",
       "      <td>93354076</td>\n",
       "      <td>chr1:93353903-93354076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22250</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93360493</td>\n",
       "      <td>93369493</td>\n",
       "      <td>DR1</td>\n",
       "      <td>93360488</td>\n",
       "      <td>93369498</td>\n",
       "      <td>chr1:93360488-93369498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35812</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154961474</td>\n",
       "      <td>154961782</td>\n",
       "      <td>PYGO2</td>\n",
       "      <td>154961469</td>\n",
       "      <td>154961787</td>\n",
       "      <td>chr1:154961469-154961787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35815</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154960977</td>\n",
       "      <td>154961026</td>\n",
       "      <td>PYGO2</td>\n",
       "      <td>154960972</td>\n",
       "      <td>154961031</td>\n",
       "      <td>chr1:154960972-154961031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499651</th>\n",
       "      <td>chrX</td>\n",
       "      <td>54440404</td>\n",
       "      <td>54440502</td>\n",
       "      <td>TSR2</td>\n",
       "      <td>54440399</td>\n",
       "      <td>54440507</td>\n",
       "      <td>chrX:54440399-54440507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499654</th>\n",
       "      <td>chrX</td>\n",
       "      <td>54440690</td>\n",
       "      <td>54440780</td>\n",
       "      <td>TSR2</td>\n",
       "      <td>54440685</td>\n",
       "      <td>54440785</td>\n",
       "      <td>chrX:54440685-54440785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499656</th>\n",
       "      <td>chrX</td>\n",
       "      <td>54443400</td>\n",
       "      <td>54443491</td>\n",
       "      <td>TSR2</td>\n",
       "      <td>54443395</td>\n",
       "      <td>54443496</td>\n",
       "      <td>chrX:54443395-54443496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499658</th>\n",
       "      <td>chrX</td>\n",
       "      <td>54444008</td>\n",
       "      <td>54444184</td>\n",
       "      <td>TSR2</td>\n",
       "      <td>54444003</td>\n",
       "      <td>54444189</td>\n",
       "      <td>chrX:54444003-54444189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499660</th>\n",
       "      <td>chrX</td>\n",
       "      <td>54444416</td>\n",
       "      <td>54448032</td>\n",
       "      <td>TSR2</td>\n",
       "      <td>54444411</td>\n",
       "      <td>54448037</td>\n",
       "      <td>chrX:54444411-54448037</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname      start        end gene_name  exon_flank_start  \\\n",
       "22245     chr1   93345907   93346865       DR1          93345902   \n",
       "22248     chr1   93353908   93354071       DR1          93353903   \n",
       "22250     chr1   93360493   93369493       DR1          93360488   \n",
       "35812     chr1  154961474  154961782     PYGO2         154961469   \n",
       "35815     chr1  154960977  154961026     PYGO2         154960972   \n",
       "...        ...        ...        ...       ...               ...   \n",
       "499651    chrX   54440404   54440502      TSR2          54440399   \n",
       "499654    chrX   54440690   54440780      TSR2          54440685   \n",
       "499656    chrX   54443400   54443491      TSR2          54443395   \n",
       "499658    chrX   54444008   54444184      TSR2          54444003   \n",
       "499660    chrX   54444416   54448032      TSR2          54444411   \n",
       "\n",
       "        exon_flank_end                    region  \n",
       "22245         93346870    chr1:93345902-93346870  \n",
       "22248         93354076    chr1:93353903-93354076  \n",
       "22250         93369498    chr1:93360488-93369498  \n",
       "35812        154961787  chr1:154961469-154961787  \n",
       "35815        154961031  chr1:154960972-154961031  \n",
       "...                ...                       ...  \n",
       "499651        54440507    chrX:54440399-54440507  \n",
       "499654        54440785    chrX:54440685-54440785  \n",
       "499656        54443496    chrX:54443395-54443496  \n",
       "499658        54444189    chrX:54444003-54444189  \n",
       "499660        54448037    chrX:54444411-54448037  \n",
       "\n",
       "[76 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz\")\n",
    "df = df.loc[(df['feature'] == 'exon') & (df['gene_name'].isin(genes))]\n",
    "df = df[['seqname', 'start', 'end', 'gene_name']]\n",
    "df['exon_flank_start'] = df['start'] - exon_flank_nt\n",
    "df['exon_flank_end'] = df['end'] + exon_flank_nt\n",
    "df['region'] = ((df['seqname'] + ':').str.cat(df['exon_flank_start'].astype(str)) + '-').str.cat(df['exon_flank_end'].astype(str))\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e66005-ff41-4b75-a443-a2ed9df136bd",
   "metadata": {},
   "source": [
    "### Run Swiss-army-knife on DNAnexus\n",
    "- get region info for each gene from block file\n",
    "- bcftools command for step 2 finished\n",
    "- bcftools command for step 3 & 4 TBD\n",
    "- list genes not included in MANE set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "14ba3303-3410-4c56-a1f4-c6201572c1be",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-Gfgf1GjJqBj9ZgK2J47KqpGp\n",
      "job-Gfgf1J0JqBj57Q9Bj6Xx5kX3\n",
      "job-Gfgf1J8JqBj5g9yvZyK3gP0z\n",
      "job-Gfgf1JjJqBj39j9QKbvk82p5\n",
      "job-Gfgf1K8JqBj8V4FPY4F8XBZ5\n",
      "job-Gfgf1PQJqBj7p3jG1y2gQj42\n",
      "job-Gfgf1PjJqBj9ZgK2J47KqpGv\n",
      "job-Gfgf1Q0JqBjFBV30Pz1GkK8g\n",
      "job-Gfgf1Q8JqBjF4p2kfb4bZ2ff\n",
      "job-Gfgf1QQJqBjBK6v14VZQj0vj\n",
      "Genes not found in MANE database:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "genes_not_found = []\n",
    "genes_found = []\n",
    "known_large_genes = [\"DSP\", \"TSC2\", \"TTN\", \"NCOA3\"]\n",
    "\n",
    "for gene in genes:\n",
    "    df_gene = df.loc[df['gene_name'] == gene]\n",
    "    if df_gene.shape[0] > 0:\n",
    "        genes_found.append(gene)\n",
    "        \n",
    "        vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "        vcf_str = ' '.join(vcf_files)\n",
    "        region_str = ','.join(df_gene['region'].to_list())\n",
    "        \n",
    "        if ((len(vcf_files) > 1) or (gene in known_large_genes)):\n",
    "            mem_level = \"mem2_ssd1_v2_x16\" # dynamically change memory level when submitting jobs on DNAnexus\n",
    "        else:\n",
    "            mem_level = \"mem1_ssd1_v2_x4\"\n",
    "        \n",
    "        # filtering variants\n",
    "        bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "        bcftools_cmd2 = \"bcftools view -Ou --max-alleles 5 -T ^\" + diff_bed\n",
    "        bcftools_cmd3 = \"bcftools +fill-tags -Ou -- -t all\"\n",
    "        bcftools_cmd4 = \"bcftools norm -Ou -m - -f \" + ref_genome \n",
    "        bcftools_cmd5 = \"bcftools view -Oz -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \"_variants.vcf.gz\"\n",
    "        \n",
    "        # extract carriers\n",
    "        bcftools_cmd6 = \"mkdir -p carrier_out\"\n",
    "        bcftools_cmd7 = \"bcftools query -i 'GT=\\\\\\\"RA\\\\\\\"|GT=\\\\\\\"AR\\\\\\\"|GT=\\\\\\\"AA\\\\\\\"' -f '%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n' \" + gene + \"_variants.vcf.gz > carrier_out/\" + gene + \".ssv\"\n",
    "        \n",
    "        # parsing command\n",
    "        bcftools_command_a = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "        bcftools_command_b = \" && \".join([bcftools_cmd6, bcftools_cmd7])\n",
    "        bcftools_command = bcftools_command_a + \" && \" + bcftools_command_b\n",
    "        \n",
    "        # parsing input\n",
    "        dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "        \n",
    "        # final dx command\n",
    "        dx_command = 'dx run swiss-army-knife --instance-type ' + mem_level + ' -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "        !{dx_command}\n",
    "    else:\n",
    "        genes_not_found.append(gene)\n",
    "            \n",
    "print('Genes not found in MANE database:')\n",
    "print(genes_not_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "adf05525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dx run swiss-army-knife --instance-type mem2_ssd1_v2_x16 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr7:6008992-6009054,chr7:6005887-6006036,chr7:6003967-6004063,chr7:6003685-6003797,chr7:6002448-6002641,chr7:5999103-5999280,chr7:5997321-5997428,chr7:5995529-5995638,chr7:5991968-5992062,chr7:5989795-5989960,chr7:5986754-5987625,chr7:5982819-5982996,chr7:5978591-5978701,chr7:5977583-5977762,chr7:5970920-5973547 ukb23157_c7_b4_v1.vcf.gz ukb23157_c7_b3_v1.vcf.gz | bcftools view -Ou --max-alleles 5 -T ^GRCh38_alldifficultregions.bed.gz | bcftools +fill-tags -Ou -- -t all | bcftools norm -Oz -m - -f GRCh38_reference_genome.fa> filtered.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/Tian_folder/Vineel_genes_test/ --tag \"Vineel_090823\" --property gene=PMS2'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dx_command_c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e454b2db-5ba6-446f-8c3c-31c164c7df92",
   "metadata": {},
   "source": [
    "### Below is for testing purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4df214f4-69b6-4396-a167-e7a0ff599e41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\n",
      "dx run swiss-army-knife --instance-type mem2_ssd1_v2_x4 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/temp_vcf_out/ --tag \"my_tag\" --property gene=APOB\n"
     ]
    }
   ],
   "source": [
    "gene = 'APOB'\n",
    "df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "vcf_str = ' '.join(vcf_files)\n",
    "region_str = ','.join(df_gene['region'].to_list())\n",
    "mem_level = len(vcf_files)\n",
    "\n",
    "bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "bcftools_cmd2 = \"bcftools view -Ou -T ^\" + diff_bed\n",
    "bcftools_cmd3 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "bcftools_cmd4 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "bcftools_cmd5 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "\n",
    "dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "dx_command = 'dx run swiss-army-knife --instance-type mem' + str(mem_level) + '_ssd1_v2_x4 -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "\n",
    "print(bcftools_command)\n",
    "print(dx_command)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4d4d5547-7ac4-4a6b-8409-130c10019c9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GPG9KY8JqBj5gK412JXv3f8G\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!{dx_command}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
