{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9a7db64-5615-462a-adc3-446da0767246",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "from subprocess import call\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8edf64c6-81e8-412c-968d-115baeaa3be2",
   "metadata": {},
   "source": [
    "### Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\n",
    "number_of_threads = 4 # number of threads used in bcftools output compression\n",
    "tag_str = 'Vineel_091823' # DNAnexus job tag\n",
    "\n",
    "project_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n",
    "\n",
    "# input vcf path (end with '/')\n",
    "dx_vcf_path = project_path + \"Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/\"\n",
    "# output vcf path (end with '/')\n",
    "dx_vcf_out_path = project_path + \"Tian_folder/Selin_genes_101223/standard_filter/\"\n",
    "# resource path (end with '/')\n",
    "dx_resource_path = project_path + \"GRCh38_resources/\"\n",
    "# difficult regions bed filename\n",
    "diff_bed = 'GRCh38_alldifficultregions.bed.gz'\n",
    "# reference genome filename\n",
    "ref_genome = 'GRCh38_reference_genome.fa' # index filename is inferred\n",
    "\n",
    "!dx mkdir -p {dx_vcf_out_path} # create gene.vcf output folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4000040-9b97-4085-a63e-b5aec384499f",
   "metadata": {},
   "source": [
    "### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2a8f096b-5b91-43aa-b546-4eaec039c30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_UKB_vcfs(df_gene, df_blk, diagnosis=False):\n",
    "    vcf_prefix = 'ukb23157_'\n",
    "    vcf_suffix = '_v1.vcf.gz'\n",
    "    \n",
    "    all_vcf_files = []\n",
    "    df_gene = df_gene.assign(vcf_files = '')\n",
    "    for index, row in df_gene.iterrows():\n",
    "        new_vcf_files = df_blk.loc[(df_blk['seqname'] == row['seqname']) & \n",
    "                                   ((df_blk['end_pos'] >= row['exon_flank_start']) & (df_blk['start_pos'] <= row['exon_flank_end'])),\n",
    "                                   'chr_blk_str'].tolist()\n",
    "        new_vcf_files = list(vcf_prefix + x + vcf_suffix for x in set(new_vcf_files))\n",
    "        df_gene.loc[index,'vcf_files'] = ','.join(new_vcf_files)\n",
    "        all_vcf_files = all_vcf_files + new_vcf_files\n",
    "    all_vcf_files = list(set(all_vcf_files))\n",
    "    \n",
    "    if diagnosis:\n",
    "        return(df_gene)\n",
    "    else:\n",
    "        return(all_vcf_files)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed62c94-9b34-42e2-90eb-67a527b04894",
   "metadata": {},
   "source": [
    "### List of gene symbols as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8fbe087c-8e0b-4b71-9f81-d9fbada151d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"FLNC\", \"TNNI3\"]\n",
    "\n",
    "# # to read from txt:\n",
    "# with open(\"resources/my_gene_file.txt\", \"r\") as gene_file:\n",
    "#     genes = gene_file.readlines()\n",
    "#     genes = [l.replace(\"\\n\", \"\") for l in genes]\n",
    "#     print(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "214b5b7e-a208-4c88-bf56-19dbcf0034bd",
   "metadata": {},
   "source": [
    "### Load pVCF block coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "62a230e4-d95d-4084-a914-01abecc3ff92",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>chr</th>\n",
       "      <th>blk</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>seqname</th>\n",
       "      <th>chr_blk_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1218130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1218131</td>\n",
       "      <td>1426969</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1426970</td>\n",
       "      <td>1758871</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1758872</td>\n",
       "      <td>2514221</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2514222</td>\n",
       "      <td>3782130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>973</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>135552245</td>\n",
       "      <td>141897932</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>974</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>141897933</td>\n",
       "      <td>152168662</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>975</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>152168663</td>\n",
       "      <td>153788223</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>976</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>153788224</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>977</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "      <td>cY_b0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>977 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ind  chr  blk  start_pos    end_pos seqname chr_blk_str\n",
       "0      1    1    0          1    1218130    chr1       c1_b0\n",
       "1      2    1    1    1218131    1426969    chr1       c1_b1\n",
       "2      3    1    2    1426970    1758871    chr1       c1_b2\n",
       "3      4    1    3    1758872    2514221    chr1       c1_b3\n",
       "4      5    1    4    2514222    3782130    chr1       c1_b4\n",
       "..   ...  ...  ...        ...        ...     ...         ...\n",
       "972  973   23   20  135552245  141897932    chrX      cX_b20\n",
       "973  974   23   21  141897933  152168662    chrX      cX_b21\n",
       "974  975   23   22  152168663  153788223    chrX      cX_b22\n",
       "975  976   23   23  153788224  156040895    chrX      cX_b23\n",
       "976  977   24    0          1   57227415    chrY       cY_b0\n",
       "\n",
       "[977 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blk = pd.read_table(\"./resources/pvcf_blocks.txt\", sep = '\\t', names = ['ind', 'chr', 'blk', 'start_pos', 'end_pos'])\n",
    "df_blk['seqname'] = 'chr' + df_blk['chr'].map(str)\n",
    "df_blk.loc[df_blk['seqname'] == 'chr23', 'seqname'] = 'chrX'\n",
    "df_blk.loc[df_blk['seqname'] == 'chr24', 'seqname'] = 'chrY'\n",
    "df_blk['chr_blk_str'] = df_blk['seqname'].str.replace('chr', 'c') + '_b' + df_blk['blk'].map(str)\n",
    "df_blk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16fc2f8f-3742-4c4e-b01e-05d35069d52d",
   "metadata": {},
   "source": [
    "### Load MANE transcript coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7c71254d-0da0-4458-92dd-3fcdf52e9b7b",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_flank_start</th>\n",
       "      <th>exon_flank_end</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>204879</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128830406</td>\n",
       "      <td>128830989</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128830401</td>\n",
       "      <td>128830994</td>\n",
       "      <td>chr7:128830401-128830994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204882</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128835326</td>\n",
       "      <td>128835574</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128835321</td>\n",
       "      <td>128835579</td>\n",
       "      <td>chr7:128835321-128835579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204884</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128837160</td>\n",
       "      <td>128837257</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128837155</td>\n",
       "      <td>128837262</td>\n",
       "      <td>chr7:128837155-128837262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204886</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128837398</td>\n",
       "      <td>128837548</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128837393</td>\n",
       "      <td>128837553</td>\n",
       "      <td>chr7:128837393-128837553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204888</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128837637</td>\n",
       "      <td>128837755</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128837632</td>\n",
       "      <td>128837760</td>\n",
       "      <td>chr7:128837632-128837760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204890</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128837987</td>\n",
       "      <td>128838064</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128837982</td>\n",
       "      <td>128838069</td>\n",
       "      <td>chr7:128837982-128838069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204892</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128838267</td>\n",
       "      <td>128838429</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128838262</td>\n",
       "      <td>128838434</td>\n",
       "      <td>chr7:128838262-128838434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204894</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128838603</td>\n",
       "      <td>128838803</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128838598</td>\n",
       "      <td>128838808</td>\n",
       "      <td>chr7:128838598-128838808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204896</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128840023</td>\n",
       "      <td>128840160</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128840018</td>\n",
       "      <td>128840165</td>\n",
       "      <td>chr7:128840018-128840165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204898</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128840548</td>\n",
       "      <td>128840674</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128840543</td>\n",
       "      <td>128840679</td>\n",
       "      <td>chr7:128840543-128840679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204900</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128840834</td>\n",
       "      <td>128840970</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128840829</td>\n",
       "      <td>128840975</td>\n",
       "      <td>chr7:128840829-128840975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204902</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128841170</td>\n",
       "      <td>128841363</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128841165</td>\n",
       "      <td>128841368</td>\n",
       "      <td>chr7:128841165-128841368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204904</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128841454</td>\n",
       "      <td>128841567</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128841449</td>\n",
       "      <td>128841572</td>\n",
       "      <td>chr7:128841449-128841572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204906</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128842231</td>\n",
       "      <td>128842374</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128842226</td>\n",
       "      <td>128842379</td>\n",
       "      <td>chr7:128842226-128842379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204908</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128842575</td>\n",
       "      <td>128842698</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128842570</td>\n",
       "      <td>128842703</td>\n",
       "      <td>chr7:128842570-128842703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204910</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128842794</td>\n",
       "      <td>128842954</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128842789</td>\n",
       "      <td>128842959</td>\n",
       "      <td>chr7:128842789-128842959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204912</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128843229</td>\n",
       "      <td>128843319</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128843224</td>\n",
       "      <td>128843324</td>\n",
       "      <td>chr7:128843224-128843324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204914</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128843408</td>\n",
       "      <td>128843577</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128843403</td>\n",
       "      <td>128843582</td>\n",
       "      <td>chr7:128843403-128843582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204916</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128843796</td>\n",
       "      <td>128843913</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128843791</td>\n",
       "      <td>128843918</td>\n",
       "      <td>chr7:128843791-128843918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204918</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128844004</td>\n",
       "      <td>128844266</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128843999</td>\n",
       "      <td>128844271</td>\n",
       "      <td>chr7:128843999-128844271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204920</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128844658</td>\n",
       "      <td>128845255</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128844653</td>\n",
       "      <td>128845260</td>\n",
       "      <td>chr7:128844653-128845260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204922</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128845990</td>\n",
       "      <td>128846163</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128845985</td>\n",
       "      <td>128846168</td>\n",
       "      <td>chr7:128845985-128846168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204924</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128846301</td>\n",
       "      <td>128846463</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128846296</td>\n",
       "      <td>128846468</td>\n",
       "      <td>chr7:128846296-128846468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204926</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128846745</td>\n",
       "      <td>128846905</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128846740</td>\n",
       "      <td>128846910</td>\n",
       "      <td>chr7:128846740-128846910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204928</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128847697</td>\n",
       "      <td>128847864</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128847692</td>\n",
       "      <td>128847869</td>\n",
       "      <td>chr7:128847692-128847869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204930</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128847945</td>\n",
       "      <td>128848068</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128847940</td>\n",
       "      <td>128848073</td>\n",
       "      <td>chr7:128847940-128848073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204932</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128848561</td>\n",
       "      <td>128848717</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128848556</td>\n",
       "      <td>128848722</td>\n",
       "      <td>chr7:128848556-128848722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204934</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128848793</td>\n",
       "      <td>128848982</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128848788</td>\n",
       "      <td>128848987</td>\n",
       "      <td>chr7:128848788-128848987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204936</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128849181</td>\n",
       "      <td>128849204</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128849176</td>\n",
       "      <td>128849209</td>\n",
       "      <td>chr7:128849176-128849209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204938</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128849331</td>\n",
       "      <td>128849578</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128849326</td>\n",
       "      <td>128849583</td>\n",
       "      <td>chr7:128849326-128849583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204940</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128849976</td>\n",
       "      <td>128850074</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128849971</td>\n",
       "      <td>128850079</td>\n",
       "      <td>chr7:128849971-128850079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204942</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128850384</td>\n",
       "      <td>128850483</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128850379</td>\n",
       "      <td>128850488</td>\n",
       "      <td>chr7:128850379-128850488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204944</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128850803</td>\n",
       "      <td>128850943</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128850798</td>\n",
       "      <td>128850948</td>\n",
       "      <td>chr7:128850798-128850948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204946</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128851232</td>\n",
       "      <td>128851360</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128851227</td>\n",
       "      <td>128851365</td>\n",
       "      <td>chr7:128851227-128851365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204948</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128851455</td>\n",
       "      <td>128851628</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128851450</td>\n",
       "      <td>128851633</td>\n",
       "      <td>chr7:128851450-128851633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204950</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128852591</td>\n",
       "      <td>128852752</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128852586</td>\n",
       "      <td>128852757</td>\n",
       "      <td>chr7:128852586-128852757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204952</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128852828</td>\n",
       "      <td>128853031</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128852823</td>\n",
       "      <td>128853036</td>\n",
       "      <td>chr7:128852823-128853036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204954</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128853469</td>\n",
       "      <td>128853621</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128853464</td>\n",
       "      <td>128853626</td>\n",
       "      <td>chr7:128853464-128853626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204956</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128853715</td>\n",
       "      <td>128853837</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128853710</td>\n",
       "      <td>128853842</td>\n",
       "      <td>chr7:128853710-128853842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204958</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128853974</td>\n",
       "      <td>128854216</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128853969</td>\n",
       "      <td>128854221</td>\n",
       "      <td>chr7:128853969-128854221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204960</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128854413</td>\n",
       "      <td>128854682</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128854408</td>\n",
       "      <td>128854687</td>\n",
       "      <td>chr7:128854408-128854687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204962</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128854775</td>\n",
       "      <td>128854912</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128854770</td>\n",
       "      <td>128854917</td>\n",
       "      <td>chr7:128854770-128854917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204964</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128855199</td>\n",
       "      <td>128855314</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128855194</td>\n",
       "      <td>128855319</td>\n",
       "      <td>chr7:128855194-128855319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204966</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128856518</td>\n",
       "      <td>128856650</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128856513</td>\n",
       "      <td>128856655</td>\n",
       "      <td>chr7:128856513-128856655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204968</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128856745</td>\n",
       "      <td>128856921</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128856740</td>\n",
       "      <td>128856926</td>\n",
       "      <td>chr7:128856740-128856926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204970</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128857118</td>\n",
       "      <td>128857336</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128857113</td>\n",
       "      <td>128857341</td>\n",
       "      <td>chr7:128857113-128857341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204972</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128858008</td>\n",
       "      <td>128858217</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128858003</td>\n",
       "      <td>128858222</td>\n",
       "      <td>chr7:128858003-128858222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204974</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128858336</td>\n",
       "      <td>128859272</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>128858331</td>\n",
       "      <td>128859277</td>\n",
       "      <td>chr7:128858331-128859277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444069</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55157579</td>\n",
       "      <td>55157732</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55157574</td>\n",
       "      <td>55157737</td>\n",
       "      <td>chr19:55157574-55157737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444072</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55157296</td>\n",
       "      <td>55157308</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55157291</td>\n",
       "      <td>55157313</td>\n",
       "      <td>chr19:55157291-55157313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444074</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55157050</td>\n",
       "      <td>55157133</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55157045</td>\n",
       "      <td>55157138</td>\n",
       "      <td>chr19:55157045-55157138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444076</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55156603</td>\n",
       "      <td>55156644</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55156598</td>\n",
       "      <td>55156649</td>\n",
       "      <td>chr19:55156598-55156649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444078</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55156201</td>\n",
       "      <td>55156332</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55156196</td>\n",
       "      <td>55156337</td>\n",
       "      <td>chr19:55156196-55156337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444080</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55154741</td>\n",
       "      <td>55154830</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55154736</td>\n",
       "      <td>55154835</td>\n",
       "      <td>chr19:55154736-55154835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444082</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55154030</td>\n",
       "      <td>55154206</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55154025</td>\n",
       "      <td>55154211</td>\n",
       "      <td>chr19:55154025-55154211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444084</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55151767</td>\n",
       "      <td>55151917</td>\n",
       "      <td>TNNI3</td>\n",
       "      <td>55151762</td>\n",
       "      <td>55151922</td>\n",
       "      <td>chr19:55151762-55151922</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname      start        end gene_name  exon_flank_start   \n",
       "204879    chr7  128830406  128830989      FLNC         128830401  \\\n",
       "204882    chr7  128835326  128835574      FLNC         128835321   \n",
       "204884    chr7  128837160  128837257      FLNC         128837155   \n",
       "204886    chr7  128837398  128837548      FLNC         128837393   \n",
       "204888    chr7  128837637  128837755      FLNC         128837632   \n",
       "204890    chr7  128837987  128838064      FLNC         128837982   \n",
       "204892    chr7  128838267  128838429      FLNC         128838262   \n",
       "204894    chr7  128838603  128838803      FLNC         128838598   \n",
       "204896    chr7  128840023  128840160      FLNC         128840018   \n",
       "204898    chr7  128840548  128840674      FLNC         128840543   \n",
       "204900    chr7  128840834  128840970      FLNC         128840829   \n",
       "204902    chr7  128841170  128841363      FLNC         128841165   \n",
       "204904    chr7  128841454  128841567      FLNC         128841449   \n",
       "204906    chr7  128842231  128842374      FLNC         128842226   \n",
       "204908    chr7  128842575  128842698      FLNC         128842570   \n",
       "204910    chr7  128842794  128842954      FLNC         128842789   \n",
       "204912    chr7  128843229  128843319      FLNC         128843224   \n",
       "204914    chr7  128843408  128843577      FLNC         128843403   \n",
       "204916    chr7  128843796  128843913      FLNC         128843791   \n",
       "204918    chr7  128844004  128844266      FLNC         128843999   \n",
       "204920    chr7  128844658  128845255      FLNC         128844653   \n",
       "204922    chr7  128845990  128846163      FLNC         128845985   \n",
       "204924    chr7  128846301  128846463      FLNC         128846296   \n",
       "204926    chr7  128846745  128846905      FLNC         128846740   \n",
       "204928    chr7  128847697  128847864      FLNC         128847692   \n",
       "204930    chr7  128847945  128848068      FLNC         128847940   \n",
       "204932    chr7  128848561  128848717      FLNC         128848556   \n",
       "204934    chr7  128848793  128848982      FLNC         128848788   \n",
       "204936    chr7  128849181  128849204      FLNC         128849176   \n",
       "204938    chr7  128849331  128849578      FLNC         128849326   \n",
       "204940    chr7  128849976  128850074      FLNC         128849971   \n",
       "204942    chr7  128850384  128850483      FLNC         128850379   \n",
       "204944    chr7  128850803  128850943      FLNC         128850798   \n",
       "204946    chr7  128851232  128851360      FLNC         128851227   \n",
       "204948    chr7  128851455  128851628      FLNC         128851450   \n",
       "204950    chr7  128852591  128852752      FLNC         128852586   \n",
       "204952    chr7  128852828  128853031      FLNC         128852823   \n",
       "204954    chr7  128853469  128853621      FLNC         128853464   \n",
       "204956    chr7  128853715  128853837      FLNC         128853710   \n",
       "204958    chr7  128853974  128854216      FLNC         128853969   \n",
       "204960    chr7  128854413  128854682      FLNC         128854408   \n",
       "204962    chr7  128854775  128854912      FLNC         128854770   \n",
       "204964    chr7  128855199  128855314      FLNC         128855194   \n",
       "204966    chr7  128856518  128856650      FLNC         128856513   \n",
       "204968    chr7  128856745  128856921      FLNC         128856740   \n",
       "204970    chr7  128857118  128857336      FLNC         128857113   \n",
       "204972    chr7  128858008  128858217      FLNC         128858003   \n",
       "204974    chr7  128858336  128859272      FLNC         128858331   \n",
       "444069   chr19   55157579   55157732     TNNI3          55157574   \n",
       "444072   chr19   55157296   55157308     TNNI3          55157291   \n",
       "444074   chr19   55157050   55157133     TNNI3          55157045   \n",
       "444076   chr19   55156603   55156644     TNNI3          55156598   \n",
       "444078   chr19   55156201   55156332     TNNI3          55156196   \n",
       "444080   chr19   55154741   55154830     TNNI3          55154736   \n",
       "444082   chr19   55154030   55154206     TNNI3          55154025   \n",
       "444084   chr19   55151767   55151917     TNNI3          55151762   \n",
       "\n",
       "        exon_flank_end                    region  \n",
       "204879       128830994  chr7:128830401-128830994  \n",
       "204882       128835579  chr7:128835321-128835579  \n",
       "204884       128837262  chr7:128837155-128837262  \n",
       "204886       128837553  chr7:128837393-128837553  \n",
       "204888       128837760  chr7:128837632-128837760  \n",
       "204890       128838069  chr7:128837982-128838069  \n",
       "204892       128838434  chr7:128838262-128838434  \n",
       "204894       128838808  chr7:128838598-128838808  \n",
       "204896       128840165  chr7:128840018-128840165  \n",
       "204898       128840679  chr7:128840543-128840679  \n",
       "204900       128840975  chr7:128840829-128840975  \n",
       "204902       128841368  chr7:128841165-128841368  \n",
       "204904       128841572  chr7:128841449-128841572  \n",
       "204906       128842379  chr7:128842226-128842379  \n",
       "204908       128842703  chr7:128842570-128842703  \n",
       "204910       128842959  chr7:128842789-128842959  \n",
       "204912       128843324  chr7:128843224-128843324  \n",
       "204914       128843582  chr7:128843403-128843582  \n",
       "204916       128843918  chr7:128843791-128843918  \n",
       "204918       128844271  chr7:128843999-128844271  \n",
       "204920       128845260  chr7:128844653-128845260  \n",
       "204922       128846168  chr7:128845985-128846168  \n",
       "204924       128846468  chr7:128846296-128846468  \n",
       "204926       128846910  chr7:128846740-128846910  \n",
       "204928       128847869  chr7:128847692-128847869  \n",
       "204930       128848073  chr7:128847940-128848073  \n",
       "204932       128848722  chr7:128848556-128848722  \n",
       "204934       128848987  chr7:128848788-128848987  \n",
       "204936       128849209  chr7:128849176-128849209  \n",
       "204938       128849583  chr7:128849326-128849583  \n",
       "204940       128850079  chr7:128849971-128850079  \n",
       "204942       128850488  chr7:128850379-128850488  \n",
       "204944       128850948  chr7:128850798-128850948  \n",
       "204946       128851365  chr7:128851227-128851365  \n",
       "204948       128851633  chr7:128851450-128851633  \n",
       "204950       128852757  chr7:128852586-128852757  \n",
       "204952       128853036  chr7:128852823-128853036  \n",
       "204954       128853626  chr7:128853464-128853626  \n",
       "204956       128853842  chr7:128853710-128853842  \n",
       "204958       128854221  chr7:128853969-128854221  \n",
       "204960       128854687  chr7:128854408-128854687  \n",
       "204962       128854917  chr7:128854770-128854917  \n",
       "204964       128855319  chr7:128855194-128855319  \n",
       "204966       128856655  chr7:128856513-128856655  \n",
       "204968       128856926  chr7:128856740-128856926  \n",
       "204970       128857341  chr7:128857113-128857341  \n",
       "204972       128858222  chr7:128858003-128858222  \n",
       "204974       128859277  chr7:128858331-128859277  \n",
       "444069        55157737   chr19:55157574-55157737  \n",
       "444072        55157313   chr19:55157291-55157313  \n",
       "444074        55157138   chr19:55157045-55157138  \n",
       "444076        55156649   chr19:55156598-55156649  \n",
       "444078        55156337   chr19:55156196-55156337  \n",
       "444080        55154835   chr19:55154736-55154835  \n",
       "444082        55154211   chr19:55154025-55154211  \n",
       "444084        55151922   chr19:55151762-55151922  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz\")\n",
    "df = df.loc[(df['feature'] == 'exon') & (df['gene_name'].isin(genes))]\n",
    "df = df[['seqname', 'start', 'end', 'gene_name']]\n",
    "df['exon_flank_start'] = df['start'] - exon_flank_nt\n",
    "df['exon_flank_end'] = df['end'] + exon_flank_nt\n",
    "df['region'] = ((df['seqname'] + ':').str.cat(df['exon_flank_start'].astype(str)) + '-').str.cat(df['exon_flank_end'].astype(str))\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01307ebf",
   "metadata": {},
   "source": [
    "#### Test vcfByGene\n",
    "- generate vcf files for each exons of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "43107bf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GZZ4JxjJqBj3bKFzjxZFxxyP\n",
      "\u001b[0mjob-GZZ4Jy0JqBj0Q6g82BV3VvV6\n",
      "\u001b[0mjob-GZZ4Jy8JqBjPJyXYpYkK2p32\n",
      "\u001b[0mjob-GZZ4JyQJqBj2k01pG9bG5Y7P\n",
      "\u001b[0mjob-GZZ4JyjJqBj4Kp07Bb7vP88Y\n",
      "\u001b[0mjob-GZZ4Jz0JqBj87VYqzJPK4qVP\n",
      "\u001b[0mjob-GZZ4Jz8JqBj2k01pG9bG5Y7Y\n",
      "\u001b[0mjob-GZZ4JzjJqBj89vgYk7678zQ2\n",
      "\u001b[0mjob-GZZ4K00JqBj4VPPbGP9b8zZZ\n",
      "\u001b[0mjob-GZZ4K08JqBj8vpXJ3vfBYq8v\n",
      "\u001b[0mjob-GZZ4K0QJqBj1ZZ7GVQgJ85x1\n",
      "\u001b[0mjob-GZZ4K0jJqBjFzJGVgJ5bxGxk\n",
      "\u001b[0mjob-GZZ4K10JqBj8Q0kV888b7Yzv\n",
      "\u001b[0mjob-GZZ4K18JqBj7xG9K3pB7g7q1\n",
      "\u001b[0mjob-GZZ4K1QJqBj9xJJvYfgFk4zj\n",
      "\u001b[0mjob-GZZ4K20JqBjByPpxk8kY0BBK\n",
      "\u001b[0mjob-GZZ4K28JqBj4JpPkXFp465YB\n",
      "\u001b[0mjob-GZZ4K2QJqBj3XFfzQp8gj8pq\n",
      "\u001b[0mjob-GZZ4K2jJqBjBq8g3Y8FgBzZj\n",
      "\u001b[0mjob-GZZ4K30JqBj1FgXqV6XG4k69\n",
      "\u001b[0mjob-GZZ4K38JqBj4X0Vq7GZGzP4Z\n",
      "\u001b[0mjob-GZZ4K3QJqBj8V8fPf3kZ6ZX5\n",
      "\u001b[0mjob-GZZ4K3jJqBjJx96VGGQ18FJ0\n",
      "\u001b[0mjob-GZZ4K48JqBj3z9vbb8P59Q8P\n",
      "\u001b[0mjob-GZZ4K4QJqBjFzJGVgJ5bxGxq\n",
      "\u001b[0mjob-GZZ4K4jJqBj0xYZ7q02JK85Y\n",
      "\u001b[0mjob-GZZ4K50JqBj3630JvYVZv692\n",
      "\u001b[0mjob-GZZ4K58JqBj0jPfXy2K1YGxQ\n",
      "\u001b[0mjob-GZZ4K5QJqBjGzPVQxqKFBjP1\n",
      "\u001b[0mjob-GZZ4K5jJqBjGz46F8yZg0Kkv\n",
      "\u001b[0mjob-GZZ4K60JqBjFzJGVgJ5bxGy0\n",
      "\u001b[0mjob-GZZ4K68JqBjPJyXYpYkK2p3B\n",
      "\u001b[0mjob-GZZ4K6QJqBjP6bk26xFZffZx\n",
      "\u001b[0mjob-GZZ4K70JqBj9xJJvYfgFk4zp\n",
      "\u001b[0mjob-GZZ4K78JqBj4JpPkXFp465Yj\n",
      "\u001b[0mjob-GZZ4K7QJqBj84J789p2Fj4Kz\n",
      "\u001b[0mjob-GZZ4K7jJqBj9xJJvYfgFk4zx\n",
      "\u001b[0mjob-GZZ4K80JqBj0635v5vVq0pBF\n",
      "\u001b[0mjob-GZZ4K88JqBj0KFg6j69ggkb9\n",
      "\u001b[0mjob-GZZ4K8QJqBjPX88G5Gp0pK7K\n",
      "\u001b[0mjob-GZZ4K8jJqBj0KFg6j69ggkbF\n",
      "\u001b[0mjob-GZZ4K90JqBj4KBXx485yF2kJ\n",
      "\u001b[0mjob-GZZ4K98JqBjJ12q5x1qvyxQ9\n",
      "\u001b[0mjob-GZZ4K9jJqBj3Xjx9jZqV6F53\n",
      "\u001b[0mjob-GZZ4KB0JqBj20ZvGQ4q8453b\n",
      "\u001b[0mjob-GZZ4KB8JqBjG445Kq8B28Q1f\n",
      "\u001b[0mjob-GZZ4KBQJqBj0jpxg25jkkKJ4\n",
      "\u001b[0mjob-GZZ4KBjJqBjGGxQvgGVy1ZFp\n",
      "\u001b[0mjob-GZZ4KF0JqBjP6bk26xFZffb1\n",
      "\u001b[0mjob-GZZ4KF8JqBj34kP446Py5YyZ\n",
      "\u001b[0mjob-GZZ4KFQJqBj0xYZ7q02JK85b\n",
      "\u001b[0mjob-GZZ4KFjJqBjP1Bk5kVBz27Kg\n",
      "\u001b[0mjob-GZZ4KG8JqBj29476V8b3Vq89\n",
      "\u001b[0mjob-GZZ4KGQJqBjGX5Kgf1x258FQ\n",
      "\u001b[0mjob-GZZ4KGjJqBjP6bk26xFZffb3\n",
      "\u001b[0mjob-GZZ4KJ0JqBj0KFg6j69ggkbV\n",
      "\u001b[0mGenes not found in MANE database:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "genes_not_found = []\n",
    "genes_found = []\n",
    "known_large_genes = [\"DSP\", \"TSC2\", \"TTN\", \"NCOA3\"]\n",
    "\n",
    "for gene in genes:\n",
    "    df_gene = df.loc[df['gene_name'] == gene]\n",
    "    if df_gene.shape[0] > 0:\n",
    "        genes_found.append(gene)\n",
    "        \n",
    "        df_gene = get_overlapping_UKB_vcfs(df_gene, df_blk, diagnosis=True)\n",
    "        df_gene['region2'] = ((df_gene['seqname'] + '_').str.cat(df['exon_flank_start'].astype(str)) + '_').str.cat(df['exon_flank_end'].astype(str))\n",
    "        \n",
    "        for index, row in df_gene.iterrows():\n",
    "            vcf_files = row['vcf_files'].split(',')\n",
    "            vcf_str = ' '.join(vcf_files)\n",
    "            region_str = row['region']\n",
    "            region_str2 = row['region2']\n",
    "\n",
    "            if ((len(vcf_files) > 1) or (gene in known_large_genes)):\n",
    "                mem_level = \"mem2_ssd1_v2_x16\" # dynamically change memory level when submitting jobs on DNAnexus\n",
    "            else:\n",
    "                mem_level = \"mem1_ssd1_v2_x4\"\n",
    "                \n",
    "#             bcftools_command = \"bcftools view -Oz -r \" + region_str + \" \" + vcf_str + \" > \" + gene + \"__\" + region_str2 + \".vcf.gz\"\n",
    "\n",
    "            # filtering variants\n",
    "            bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "            bcftools_cmd2 = \"bcftools view -Ou --max-alleles 5 -T ^\" + diff_bed\n",
    "            bcftools_cmd3 = \"bcftools +fill-tags -Ou -- -t all\"\n",
    "            bcftools_cmd4 = \"bcftools norm -Ou -m - -f \" + ref_genome \n",
    "#             bcftools_cmd5 = \"bcftools view -Oz -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \"_variants.vcf.gz\"\n",
    "            bcftools_cmd5 = \"bcftools view -Oz -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \"__\" + region_str2 + \".vcf.gz\"\n",
    "\n",
    "#             # extract carriers\n",
    "#             bcftools_cmd6 = \"mkdir -p carrier_out\"\n",
    "#             bcftools_cmd7 = \"bcftools query -i 'GT=\\\\\\\"RA\\\\\\\"|GT=\\\\\\\"AR\\\\\\\"|GT=\\\\\\\"AA\\\\\\\"' -f '%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n' \" + gene + \"_variants.vcf.gz > carrier_out/\" + gene + \".ssv\"\n",
    "\n",
    "            # parsing command\n",
    "            bcftools_command_a = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "#             bcftools_command_b = \" && \".join([bcftools_cmd6, bcftools_cmd7])\n",
    "#             bcftools_command = bcftools_command_a + \" && \" + bcftools_command_b\n",
    "            \n",
    "            bcftools_command = bcftools_command_a\n",
    "\n",
    "            # parsing input\n",
    "            dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "            dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "            dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "            dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "            # final dx command\n",
    "            dx_command = 'dx run swiss-army-knife --instance-type ' + mem_level + ' -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "            !{dx_command}\n",
    "    else:\n",
    "        genes_not_found.append(gene)\n",
    "            \n",
    "print('Genes not found in MANE database:')\n",
    "print(genes_not_found)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e66005-ff41-4b75-a443-a2ed9df136bd",
   "metadata": {},
   "source": [
    "### Run Swiss-army-knife on DNAnexus\n",
    "- get region info for each gene from block file\n",
    "- bcftools command for step 2 finished\n",
    "- bcftools command for step 3 & 4 TBD\n",
    "- list genes not included in MANE set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "14ba3303-3410-4c56-a1f4-c6201572c1be",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GZ3xJ0QJqBjKj6ZkfVqP4yk3\n",
      "\u001b[0mjob-GZ3xJ0jJqBj9Fxj19vQpgPVQ\n",
      "\u001b[0mjob-GZ3xJ18JqBj1Xq6p0xQ2qz8z\n",
      "\u001b[0mjob-GZ3xJ1QJqBj2yZFby8Bvj6P6\n",
      "\u001b[0mjob-GZ3xJ1jJqBjK9XFZ0VXp0vg9\n",
      "\u001b[0mjob-GZ3xJ28JqBj6pfz1fqPkfg88\n",
      "\u001b[0mjob-GZ3xJ2QJqBj3Qj53V4ZJVBZb\n",
      "\u001b[0mGenes not found in MANE database:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "genes_not_found = []\n",
    "genes_found = []\n",
    "known_large_genes = [\"DSP\", \"TSC2\", \"TTN\", \"NCOA3\"]\n",
    "\n",
    "for gene in genes:\n",
    "    df_gene = df.loc[df['gene_name'] == gene]\n",
    "    if df_gene.shape[0] > 0:\n",
    "        genes_found.append(gene)\n",
    "        \n",
    "        vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "        vcf_str = ' '.join(vcf_files)\n",
    "        region_str = ','.join(df_gene['region'].to_list())\n",
    "        \n",
    "        if ((len(vcf_files) > 1) or (gene in known_large_genes)):\n",
    "            mem_level = \"mem2_ssd1_v2_x16\" # dynamically change memory level when submitting jobs on DNAnexus\n",
    "        else:\n",
    "            mem_level = \"mem1_ssd1_v2_x4\"\n",
    "        \n",
    "        # filtering variants\n",
    "        bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "        bcftools_cmd2 = \"bcftools view -Ou --max-alleles 5 -T ^\" + diff_bed\n",
    "        bcftools_cmd3 = \"bcftools +fill-tags -Ou -- -t all\"\n",
    "        bcftools_cmd4 = \"bcftools norm -Ou -m - -f \" + ref_genome \n",
    "        bcftools_cmd5 = \"bcftools view -Oz -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \"_variants.vcf.gz\"\n",
    "        \n",
    "        # extract carriers\n",
    "        bcftools_cmd6 = \"mkdir -p carrier_out\"\n",
    "        bcftools_cmd7 = \"bcftools query -i 'GT=\\\\\\\"RA\\\\\\\"|GT=\\\\\\\"AR\\\\\\\"|GT=\\\\\\\"AA\\\\\\\"' -f '%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n' \" + gene + \"_variants.vcf.gz > carrier_out/\" + gene + \".ssv\"\n",
    "        \n",
    "        # parsing command\n",
    "        bcftools_command_a = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "        bcftools_command_b = \" && \".join([bcftools_cmd6, bcftools_cmd7])\n",
    "        bcftools_command = bcftools_command_a + \" && \" + bcftools_command_b\n",
    "        \n",
    "        # parsing input\n",
    "        dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "        \n",
    "        # final dx command\n",
    "        dx_command = 'dx run swiss-army-knife --instance-type ' + mem_level + ' -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "        !{dx_command}\n",
    "    else:\n",
    "        genes_not_found.append(gene)\n",
    "            \n",
    "print('Genes not found in MANE database:')\n",
    "print(genes_not_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "adf05525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dx run swiss-army-knife --instance-type mem2_ssd1_v2_x16 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\\\ sequences/Population\\\\ level\\\\ exome\\\\ OQFE\\\\ variants,\\\\ pVCF\\\\ format\\\\ -\\\\ final\\\\ release/ukb23157_c7_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr7:6008992-6009054,chr7:6005887-6006036,chr7:6003967-6004063,chr7:6003685-6003797,chr7:6002448-6002641,chr7:5999103-5999280,chr7:5997321-5997428,chr7:5995529-5995638,chr7:5991968-5992062,chr7:5989795-5989960,chr7:5986754-5987625,chr7:5982819-5982996,chr7:5978591-5978701,chr7:5977583-5977762,chr7:5970920-5973547 ukb23157_c7_b4_v1.vcf.gz ukb23157_c7_b3_v1.vcf.gz | bcftools view -Ou --max-alleles 5 -T ^GRCh38_alldifficultregions.bed.gz | bcftools +fill-tags -Ou -- -t all | bcftools norm -Oz -m - -f GRCh38_reference_genome.fa> filtered.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/Tian_folder/Vineel_genes_test/ --tag \"Vineel_090823\" --property gene=PMS2'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dx_command_c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e454b2db-5ba6-446f-8c3c-31c164c7df92",
   "metadata": {},
   "source": [
    "### Below is for testing purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4df214f4-69b6-4396-a167-e7a0ff599e41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\n",
      "dx run swiss-army-knife --instance-type mem2_ssd1_v2_x4 -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b4_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c2_b3_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools concat -Ou -a -r chr2:21043859-21044078,chr2:21043508-21043556,chr2:21042356-21042481,chr2:21040933-21041088,chr2:21037953-21038116,chr2:21037095-21037260,chr2:21035579-21035713,chr2:21034811-21034906,chr2:21033294-21033523,chr2:21032349-21032586,chr2:21029893-21030020,chr2:21029634-21029790,chr2:21028322-21028543,chr2:21027823-21028070,chr2:21026783-21026969,chr2:21024928-21025129,chr2:21023520-21023697,chr2:21022826-21023047,chr2:21019718-21019910,chr2:21018987-21019118,chr2:21016434-21016654,chr2:21015365-21015550,chr2:21015068-21015265,chr2:21014443-21014598,chr2:21013155-21013538,chr2:21005075-21012656,chr2:21004556-21004680,chr2:21004264-21004457,chr2:21001424-21003339 ukb23157_c2_b3_v1.vcf.gz ukb23157_c2_b4_v1.vcf.gz | bcftools view -Ou -T ^GRCh38_alldifficultregions.bed.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > APOB.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/temp_vcf_out/ --tag \"my_tag\" --property gene=APOB\n"
     ]
    }
   ],
   "source": [
    "gene = 'APOB'\n",
    "df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "vcf_str = ' '.join(vcf_files)\n",
    "region_str = ','.join(df_gene['region'].to_list())\n",
    "mem_level = len(vcf_files)\n",
    "\n",
    "bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n",
    "bcftools_cmd2 = \"bcftools view -Ou -T ^\" + diff_bed\n",
    "bcftools_cmd3 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "bcftools_cmd4 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "bcftools_cmd5 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n",
    "\n",
    "dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "dx_command = 'dx run swiss-army-knife --instance-type mem' + str(mem_level) + '_ssd1_v2_x4 -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "\n",
    "print(bcftools_command)\n",
    "print(dx_command)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4d4d5547-7ac4-4a6b-8409-130c10019c9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GPG9KY8JqBj5gK412JXv3f8G\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!{dx_command}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
