{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "a9a7db64-5615-462a-adc3-446da0767246",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "from subprocess import call\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8edf64c6-81e8-412c-968d-115baeaa3be2",
   "metadata": {},
   "source": [
    "### Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\n",
    "number_of_threads = 4 # number of threads used in bcftools output compression\n",
    "tag_str = 'my_tag' # DNAnexus job tag\n",
    "\n",
    "project_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n",
    "\n",
    "# input vcf path (end with '/')\n",
    "dx_vcf_path = project_path + \"Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/\"\n",
    "# output vcf path (end with '/')\n",
    "dx_vcf_out_path = project_path + \"temp_vcf_out/\"\n",
    "# resource path (end with '/')\n",
    "dx_resource_path = project_path + \"GRCh38_resources/\"\n",
    "# difficult regions bed filename\n",
    "diff_bed = 'GRCh38_alldifficultregions.bed.gz'\n",
    "# reference genome filename\n",
    "ref_genome = 'GRCh38_reference_genome.fa' # index filename is inferred\n",
    "\n",
    "!dx mkdir -p {dx_vcf_out_path} # create gene.vcf output folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4000040-9b97-4085-a63e-b5aec384499f",
   "metadata": {},
   "source": [
    "### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "2a8f096b-5b91-43aa-b546-4eaec039c30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlapping_UKB_vcfs(df_gene, df_blk):\n",
    "    vcf_prefix = 'ukb23157_'\n",
    "    vcf_suffix = '_v1.vcf.gz'\n",
    "    \n",
    "    vcf_files = []\n",
    "    for i in range(df_gene.shape[0]):\n",
    "        vcf_files = vcf_files + df_blk.loc[(df_blk['seqname'] == df_gene[i,'seqname']) & \n",
    "                                           # when the gene is completely contained inside the block\n",
    "                                           (((df_blk['start_pos'] <= df_gene[i,'exon_flank_start']) & (df_blk['end_pos'] >= df_gene[i,'exon_flank_end'])) |\n",
    "                                            # when the block is completely contained inside the gene\n",
    "                                            ((df_blk['start_pos'] >= df_gene[i,'exon_flank_start']) & (df_blk['end_pos'] <= df_gene[i,'exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the start of the block\n",
    "                                            (df_blk['start_pos'].between(df_gene[i,'exon_flank_start'], df_gene[i,'exon_flank_end'])) |\n",
    "                                            # when the gene overlaps with the end of the block\n",
    "                                            (df_blk['end_pos'].between(df_gene[i,'exon_flank_start'], df_gene[i,'exon_flank_end']))),\n",
    "                                           'chr_blk_str'].tolist()\n",
    "    vcf_files = set(vcf_prefix + x + vcf_suffix for x in set(vcf_files))\n",
    "    return(vcf_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed62c94-9b34-42e2-90eb-67a527b04894",
   "metadata": {},
   "source": [
    "### List of gene symbols as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "8fbe087c-8e0b-4b71-9f81-d9fbada151d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['BRCA1', 'TWNK', 'TENT2', 'PRR26', 'AFDN', 'UPK3B', 'SELENOS', 'DENND11', 'FAKE_GENE']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "214b5b7e-a208-4c88-bf56-19dbcf0034bd",
   "metadata": {},
   "source": [
    "### Load pVCF block coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "62a230e4-d95d-4084-a914-01abecc3ff92",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>chr</th>\n",
       "      <th>blk</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>seqname</th>\n",
       "      <th>chr_blk_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1218130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1218131</td>\n",
       "      <td>1426969</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1426970</td>\n",
       "      <td>1758871</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1758872</td>\n",
       "      <td>2514221</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2514222</td>\n",
       "      <td>3782130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c1_b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>973</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>135552245</td>\n",
       "      <td>141897932</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>974</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>141897933</td>\n",
       "      <td>152168662</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>975</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>152168663</td>\n",
       "      <td>153788223</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>976</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>153788224</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "      <td>cX_b23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>977</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "      <td>cY_b0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>977 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ind  chr  blk  start_pos    end_pos seqname chr_blk_str\n",
       "0      1    1    0          1    1218130    chr1       c1_b0\n",
       "1      2    1    1    1218131    1426969    chr1       c1_b1\n",
       "2      3    1    2    1426970    1758871    chr1       c1_b2\n",
       "3      4    1    3    1758872    2514221    chr1       c1_b3\n",
       "4      5    1    4    2514222    3782130    chr1       c1_b4\n",
       "..   ...  ...  ...        ...        ...     ...         ...\n",
       "972  973   23   20  135552245  141897932    chrX      cX_b20\n",
       "973  974   23   21  141897933  152168662    chrX      cX_b21\n",
       "974  975   23   22  152168663  153788223    chrX      cX_b22\n",
       "975  976   23   23  153788224  156040895    chrX      cX_b23\n",
       "976  977   24    0          1   57227415    chrY       cY_b0\n",
       "\n",
       "[977 rows x 7 columns]"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blk = pd.read_table(\"./resources/pvcf_blocks.txt\", sep = '\\t', names = ['ind', 'chr', 'blk', 'start_pos', 'end_pos'])\n",
    "df_blk['seqname'] = 'chr' + df_blk['chr'].map(str)\n",
    "df_blk.loc[df_blk['seqname'] == 'chr23', 'seqname'] = 'chrX'\n",
    "df_blk.loc[df_blk['seqname'] == 'chr24', 'seqname'] = 'chrY'\n",
    "df_blk['chr_blk_str'] = df_blk['seqname'].str.replace('chr', 'c') + '_b' + df_blk['blk'].map(str)\n",
    "df_blk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16fc2f8f-3742-4c4e-b01e-05d35069d52d",
   "metadata": {},
   "source": [
    "### Load MANE transcript coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "7c71254d-0da0-4458-92dd-3fcdf52e9b7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Polars found a filename. Ensure you pass a path to the file instead of a python file object when possible for best performance.INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'transcript_id', 'transcript_type', 'transcript_name', 'tag', 'protein_id', 'db_xref', 'exon_number', 'exon_id']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        white-space: pre;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-top: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-bottom: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        line-height: 95%;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "<small>shape: (98, 7)</small>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>\n",
       "seqname\n",
       "</th>\n",
       "<th>\n",
       "start\n",
       "</th>\n",
       "<th>\n",
       "end\n",
       "</th>\n",
       "<th>\n",
       "gene_name\n",
       "</th>\n",
       "<th>\n",
       "exon_flank_start\n",
       "</th>\n",
       "<th>\n",
       "exon_flank_end\n",
       "</th>\n",
       "<th>\n",
       "region\n",
       "</th>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "cat\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79612441\n",
       "</td>\n",
       "<td>\n",
       "79613075\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79612436\n",
       "</td>\n",
       "<td>\n",
       "79613080\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79612436-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79619612\n",
       "</td>\n",
       "<td>\n",
       "79619785\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79619607\n",
       "</td>\n",
       "<td>\n",
       "79619790\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79619607-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79619994\n",
       "</td>\n",
       "<td>\n",
       "79620083\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79619989\n",
       "</td>\n",
       "<td>\n",
       "79620088\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79619989-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79623252\n",
       "</td>\n",
       "<td>\n",
       "79623489\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79623247\n",
       "</td>\n",
       "<td>\n",
       "79623494\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79623247-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79640851\n",
       "</td>\n",
       "<td>\n",
       "79640965\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79640846\n",
       "</td>\n",
       "<td>\n",
       "79640970\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79640846-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79641105\n",
       "</td>\n",
       "<td>\n",
       "79641196\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79641100\n",
       "</td>\n",
       "<td>\n",
       "79641201\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79641100-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79642832\n",
       "</td>\n",
       "<td>\n",
       "79642910\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79642827\n",
       "</td>\n",
       "<td>\n",
       "79642915\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79642827-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79645123\n",
       "</td>\n",
       "<td>\n",
       "79645192\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79645118\n",
       "</td>\n",
       "<td>\n",
       "79645197\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79645118-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79648617\n",
       "</td>\n",
       "<td>\n",
       "79648693\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79648612\n",
       "</td>\n",
       "<td>\n",
       "79648698\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79648612-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79649062\n",
       "</td>\n",
       "<td>\n",
       "79649190\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79649057\n",
       "</td>\n",
       "<td>\n",
       "79649195\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79649057-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79656958\n",
       "</td>\n",
       "<td>\n",
       "79657001\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79656953\n",
       "</td>\n",
       "<td>\n",
       "79657006\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79656953-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr5&quot;\n",
       "</td>\n",
       "<td>\n",
       "79668892\n",
       "</td>\n",
       "<td>\n",
       "79669028\n",
       "</td>\n",
       "<td>\n",
       "&quot;TENT2&quot;\n",
       "</td>\n",
       "<td>\n",
       "79668887\n",
       "</td>\n",
       "<td>\n",
       "79669033\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr5:79668887-...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43082404\n",
       "</td>\n",
       "<td>\n",
       "43082575\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43082399\n",
       "</td>\n",
       "<td>\n",
       "43082580\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43082399...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43076488\n",
       "</td>\n",
       "<td>\n",
       "43076614\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43076483\n",
       "</td>\n",
       "<td>\n",
       "43076619\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43076483...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43074331\n",
       "</td>\n",
       "<td>\n",
       "43074521\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43074326\n",
       "</td>\n",
       "<td>\n",
       "43074526\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43074326...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43070928\n",
       "</td>\n",
       "<td>\n",
       "43071238\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43070923\n",
       "</td>\n",
       "<td>\n",
       "43071243\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43070923...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43067608\n",
       "</td>\n",
       "<td>\n",
       "43067695\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43067603\n",
       "</td>\n",
       "<td>\n",
       "43067700\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43067603...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063874\n",
       "</td>\n",
       "<td>\n",
       "43063951\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063869\n",
       "</td>\n",
       "<td>\n",
       "43063956\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43063869...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063333\n",
       "</td>\n",
       "<td>\n",
       "43063373\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43063328\n",
       "</td>\n",
       "<td>\n",
       "43063378\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43063328...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43057052\n",
       "</td>\n",
       "<td>\n",
       "43057135\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43057047\n",
       "</td>\n",
       "<td>\n",
       "43057140\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43057047...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43051063\n",
       "</td>\n",
       "<td>\n",
       "43051117\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43051058\n",
       "</td>\n",
       "<td>\n",
       "43051122\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43051058...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43049121\n",
       "</td>\n",
       "<td>\n",
       "43049194\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43049116\n",
       "</td>\n",
       "<td>\n",
       "43049199\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43049116...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43047643\n",
       "</td>\n",
       "<td>\n",
       "43047703\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43047638\n",
       "</td>\n",
       "<td>\n",
       "43047708\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43047638...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "&quot;chr17&quot;\n",
       "</td>\n",
       "<td>\n",
       "43044295\n",
       "</td>\n",
       "<td>\n",
       "43045802\n",
       "</td>\n",
       "<td>\n",
       "&quot;BRCA1&quot;\n",
       "</td>\n",
       "<td>\n",
       "43044290\n",
       "</td>\n",
       "<td>\n",
       "43045807\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr17:43044290...\n",
       "</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "shape: (98, 7)\n",
       "┌─────────┬──────────┬──────────┬───────────┬──────────────────┬────────────────┬──────────────────┐\n",
       "│ seqname ┆ start    ┆ end      ┆ gene_name ┆ exon_flank_start ┆ exon_flank_end ┆ region           │\n",
       "│ ---     ┆ ---      ┆ ---      ┆ ---       ┆ ---              ┆ ---            ┆ ---              │\n",
       "│ cat     ┆ i64      ┆ i64      ┆ str       ┆ i64              ┆ i64            ┆ str              │\n",
       "╞═════════╪══════════╪══════════╪═══════════╪══════════════════╪════════════════╪══════════════════╡\n",
       "│ chr5    ┆ 79612441 ┆ 79613075 ┆ TENT2     ┆ 79612436         ┆ 79613080       ┆ chr5:79612436-79 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 613080           │\n",
       "│ chr5    ┆ 79619612 ┆ 79619785 ┆ TENT2     ┆ 79619607         ┆ 79619790       ┆ chr5:79619607-79 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 619790           │\n",
       "│ chr5    ┆ 79619994 ┆ 79620083 ┆ TENT2     ┆ 79619989         ┆ 79620088       ┆ chr5:79619989-79 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 620088           │\n",
       "│ chr5    ┆ 79623252 ┆ 79623489 ┆ TENT2     ┆ 79623247         ┆ 79623494       ┆ chr5:79623247-79 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 623494           │\n",
       "│ ...     ┆ ...      ┆ ...      ┆ ...       ┆ ...              ┆ ...            ┆ ...              │\n",
       "│ chr17   ┆ 43051063 ┆ 43051117 ┆ BRCA1     ┆ 43051058         ┆ 43051122       ┆ chr17:43051058-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3051122          │\n",
       "│ chr17   ┆ 43049121 ┆ 43049194 ┆ BRCA1     ┆ 43049116         ┆ 43049199       ┆ chr17:43049116-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3049199          │\n",
       "│ chr17   ┆ 43047643 ┆ 43047703 ┆ BRCA1     ┆ 43047638         ┆ 43047708       ┆ chr17:43047638-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3047708          │\n",
       "│ chr17   ┆ 43044295 ┆ 43045802 ┆ BRCA1     ┆ 43044290         ┆ 43045807       ┆ chr17:43044290-4 │\n",
       "│         ┆          ┆          ┆           ┆                  ┆                ┆ 3045807          │\n",
       "└─────────┴──────────┴──────────┴───────────┴──────────────────┴────────────────┴──────────────────┘"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = read_gtf(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.gtf.gz\")\n",
    "df = df.filter(pl.col('feature') == 'exon')\n",
    "df = df.filter(pl.col('gene_name').is_in(genes))\n",
    "df = df.select(pl.col(['seqname', 'start', 'end', 'gene_name']))\n",
    "df = df.with_columns([(pl.col('start') - exon_flank_nt).alias('exon_flank_start'), \n",
    "                      (pl.col('end') + exon_flank_nt).alias('exon_flank_end')])\n",
    "df = df.with_column((pl.col('seqname')+':'+pl.col('exon_flank_start')+'-'+pl.col('exon_flank_end')).alias('region'))\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e66005-ff41-4b75-a443-a2ed9df136bd",
   "metadata": {},
   "source": [
    "### Run Swiss-army-knife on DNAnexus\n",
    "- get region info for each gene from block file\n",
    "- bcftools command for step 2 finished\n",
    "- bcftools command for step 3 & 4 TBD\n",
    "- list genes not included in MANE set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "14ba3303-3410-4c56-a1f4-c6201572c1be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GP4q1GjJqBj12z5y2x15Gf90\n",
      "\u001b[0mjob-GP4q1J0JqBj4fF5K657Zx6Yy\n",
      "\u001b[0mjob-GP4q1J8JqBj0X1qg5v8G25xz\n",
      "\u001b[0mjob-GP4q1JQJqBjK2K3x7Bb6QGb5\n",
      "\u001b[0mjob-GP4q1K0JqBj7Vj105PBpzf4Q\n",
      "\u001b[0mjob-GP4q1K8JqBjPVgxF8b9FV92j\n",
      "\u001b[0mjob-GP4q1KQJqBj13ZG95vBy5b8k\n",
      "\u001b[0mGenes not found in MANE database:\n",
      "['PRR26', 'FAKE_GENE']\n"
     ]
    }
   ],
   "source": [
    "genes_not_found = []\n",
    "genes_found = []\n",
    "\n",
    "for gene in genes:\n",
    "    df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "    if df_gene.shape[0] > 0:\n",
    "        genes_found.append(gene)\n",
    "        \n",
    "        vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "        vcf_str = ' '.join(vcf_files)\n",
    "        region_str = ','.join(df_gene['region'].to_list())\n",
    "\n",
    "        bcftools_cmd1 = \"bcftools view -Ou -r \" + region_str + \" -T ^\" + diff_bed + \" \" + vcf_str\n",
    "        bcftools_cmd2 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "        bcftools_cmd3 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "        bcftools_cmd4 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "        bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4])\n",
    "\n",
    "        dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "        dx_command = 'dx run swiss-army-knife -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "        !{dx_command}\n",
    "    else:\n",
    "        genes_not_found.append(gene)\n",
    "            \n",
    "print('Genes not found in MANE database:')\n",
    "print(genes_not_found)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e454b2db-5ba6-446f-8c3c-31c164c7df92",
   "metadata": {},
   "source": [
    "### Below is for testing purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "4df214f4-69b6-4396-a167-e7a0ff599e41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dx run swiss-army-knife -y --brief -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c17_b29_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c17_b30_v1.vcf.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c17_b29_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/ukb23157_c17_b30_v1.vcf.gz.tbi\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_alldifficultregions.bed.gz\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa\" -iin=\"project-GGy3Bb0JqBj7zfxY8v4by61X:/GRCh38_resources/GRCh38_reference_genome.fa.fai\" -icmd=\"bcftools view -Ou -r chr17:43125266-43125369,chr17:43124012-43124120,chr17:43115721-43115784,chr17:43106451-43106538,chr17:43104863-43104961,chr17:43104117-43104266,chr17:43099770-43099885,chr17:43097239-43097294,chr17:43095841-43095927,chr17:43091430-43094865,chr17:43090939-43091037,chr17:43082399-43082580,chr17:43076483-43076619,chr17:43074326-43074526,chr17:43070923-43071243,chr17:43067603-43067700,chr17:43063869-43063956,chr17:43063328-43063378,chr17:43057047-43057140,chr17:43051058-43051122,chr17:43049116-43049199,chr17:43047638-43047708,chr17:43044290-43045807 -T ^GRCh38_alldifficultregions.bed.gz ukb23157_c17_b30_v1.vcf.gz ukb23157_c17_b29_v1.vcf.gz | bcftools norm -Ou -m - -f GRCh38_reference_genome.fa | bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT' | bcftools view -Oz --threads 4 -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\"mis\\\")> 0.9' > BRCA1.vcf.gz\" --destination project-GGy3Bb0JqBj7zfxY8v4by61X:/temp_vcf_out/ --tag \"my_tag\" --property gene=BRCA1\n"
     ]
    }
   ],
   "source": [
    "gene = 'BRCA1'\n",
    "df_gene = df.filter(pl.col('gene_name') == gene)\n",
    "vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n",
    "vcf_str = ' '.join(vcf_files)\n",
    "region_str = ','.join(df_gene['region'].to_list())\n",
    "\n",
    "bcftools_cmd1 = \"bcftools view -Ou -r \" + region_str + \" -T ^\" + diff_bed + \" \" + vcf_str\n",
    "bcftools_cmd2 = \"bcftools norm -Ou -m - -f \" + ref_genome\n",
    "bcftools_cmd3 = \"bcftools annotate -Ou --set-id '%CHROM\\_%POS\\_%REF\\_%ALT'\"\n",
    "bcftools_cmd4 = \"bcftools view -Oz --threads \" + str(number_of_threads) + \" -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \".vcf.gz\"\n",
    "bcftools_command = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4])\n",
    "\n",
    "dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n",
    "dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n",
    "\n",
    "dx_command = 'dx run swiss-army-knife -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n",
    "\n",
    "\n",
    "print(dx_command)\n",
    "\n",
    "\n",
    "# dx mkdir -p $VCF_OUT_PATH\n",
    "# dx run swiss-army-knife -iin=\"$VCF_PATH/${VCF_PREFIX}.vcf.gz\" -iin=\"$VCF_PATH/${VCF_PREFIX}.vcf.gz.tbi\" -iin=\"$BED_FILE\" -icmd=\"$CMD_STRING\" --destination $VCF_OUT_PATH -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "id": "4d4d5547-7ac4-4a6b-8409-130c10019c9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "job-GP4jb6QJqBjP39pF64PJJxG2\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!{dx_command}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "8e1ce33d-4d5c-47fc-b6cb-1174db16145a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ukb23157_c17_b30_v1.vcf.gz ukb23157_c17_b29_v1.vcf.gz'"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dx_input_str = ' '.join(set('-iin=' + dx_vcf_path + '/' + x for x in vcf_files))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
